---
title: Human gene expression
bibliography: bibliography.bibtex
---

```{r init, echo = FALSE, message = FALSE}
library(knitr)
if (! is.null(current_input())) { # if knitr is being used
  knitr::read_chunk("settings.R")
} else {
  source("settings.R")
}
```

## Parameters

```{r rendering_settings, echo = FALSE, warning=FALSE, message=FALSE}
```

```{r io_settings}
```

```{r}
min_p_value <- 0.05
```


## Introduction



## Dependencies

This analysis requires Bioconductor packages to work. 
These packages are not available on CRAN and must be installed using the following code. 
The code below is not run during the compilation of this page, but is provided so others can quickly find the right software to replicate this analysis. 

```{r eval = FALSE}
source("https://bioconductor.org/biocLite.R")
source("http://bioconductor.org/workflows.R")
```

```{r eval = FALSE}
workflowInstall("rnaseqGene")
biocLite("GO.db")
biocLite("org.Hs.eg.db")
biocLite("airway")
biocLite("DESeq2")
```



## Read in data

### Starting from SummarizedExperiment

http://www.bioconductor.org/help/workflows/rnaseqGene/


```{r, message=FALSE, warning=FALSE}
library(rnaseqGene)
library("airway")
data("airway")
se <- airway
```

```{r}
se$dex <- relevel(se$dex, "untrt")
se$dex
```

```{r}
library("DESeq2")
dds <- DESeqDataSet(se, design = ~ cell + dex)
```

### Pre-filtering the dataset

```{r}
nrow(dds)
dds <- dds[ rowSums(counts(dds)) > 1, ]
nrow(dds)
```

### The rlog transformation

```{r}
rld <- rlog(dds, blind=FALSE)
head(assay(rld), 3)
```


### Differential expression analysis

```{r}
dds <- estimateSizeFactors(dds)
dds <- DESeq(dds)
(res <- results(dds))
```




### Converting ENSEMBL IDs to GO IDs


```{r}
library(GO.db)
library(org.Hs.eg.db)
res$go_id <- mapIds(org.Hs.eg.db,
                    keys=rownames(res),
                    column="GO",
                    keytype="ENSEMBL",
                    multiVals="first")
```


### Fitlering results 

Remove genes with no GO annotation 

```{r eval = TRUE}
nrow(res)
res <- res[!is.na(res$go_id), ]
res <- res[res$go_id %in% keys(org.Hs.eg.db, keytype = "GO"), ]
nrow(res)
```

Remove insignificant genes

```{r eval = FALSE}
nrow(res)
res <- res[(! is.na(res$padj)) & res$padj <= 0.05, ]
nrow(res)
```

Remove genes with small changes

```{r eval = FALSE}
nrow(res)
res <-  res[abs(res$log2FoldChange) >= 0.5, ]
nrow(res)
```


### Getting classificaiton

```{r}
term_class <- function(x, current = x, all_paths = TRUE, type = GOCCPARENTS, verbose = TRUE, 
                       valid_relationships = c("is_a")) {
  # Get immediate children of current taxon
  parents = tryCatch({
    possible_parents <- as.list(type[x[1]])[[1]]
    if (! is.null(valid_relationships)) {
      possible_parents <- possible_parents[names(possible_parents) %in% valid_relationships]
    }
    names(AnnotationDbi::Term(possible_parents))
  }, error = function(e) {
    c()
  })
  
  # only go down one path if desired
  if (! all_paths) {
    parents <- parents[1]
  }
  parents <- parents[parents != "all"]
  
  if (is.null(parents)) {
    return(c())
  } else if (length(parents) == 0) {
    return(paste0(collapse = "|", AnnotationDbi::Term(x)))
  } else {
    next_x <- lapply(parents, function(y) c(y, x))
    
    # Run this function on them to get their output
    child_output <- lapply(next_x, term_class, all_paths = all_paths, type = type)
    output <- unlist(child_output, recursive = FALSE)
    
    return(output)
  }
}

```



```{r cache = TRUE}
cc_class <- lapply(res$go_id, term_class, all_paths = FALSE, type = GOCCPARENTS)
mf_class <- lapply(res$go_id, term_class, all_paths = FALSE, type = GOMFPARENTS)
bp_class <- lapply(res$go_id, term_class, all_paths = FALSE, type = GOBPPARENTS)
```

```{r echo = FALSE, eval = FALSE}
save.image(file = "09--gene_expression--workspace.RData")
# load("09--gene_expression--workspace.RData")
```

### Cellular component

```{r  cache = TRUE}
cc_res <- res[rep(1:nrow(res), sapply(cc_class, length)), ]
cc_res$class <- unlist(cc_class)
library(metacoder)
obj <- parse_tax_data(as.data.frame(cc_res), class_cols = "class", class_sep = "|")
obj$funcs <- c(obj$funcs,
               change = function(x, subset = NULL) {
                 vapply(obs(x, "tax_data"),
                        function(i) {
                          obs_change <- obj$data$tax_data[i, ]$log2FoldChange[obj$data$tax_data[i, ]$padj <= min_p_value]
                          mean(obs_change, na.rm = TRUE)
                        },
                        numeric(1))
               },
               num_changed = function(x, subset = NULL) {
                 vapply(obs(x, "tax_data"),
                        function(i) {
                          sum(obj$data$tax_data[i, ]$padj <= min_p_value, na.rm = TRUE)
                        },
                        numeric(1))
               })
```


```{r cache = TRUE}
set.seed(3)
obj %>%
  filter_taxa(num_changed > 0) %>%
  filter_taxa(n_supertaxa <= 4) %>%
  # filter_taxa(n_supertaxa >= 1) %>% 
  # filter_taxa(nchar(taxon_names) <= 40) %>%
  heat_tree(node_label = ifelse(nchar(taxon_names) <= 50, taxon_names, ""),
            node_size = num_changed,
            # node_size_trans = "log10",
            node_size_range = c(0.01, 0.03),
            # node_label_size_trans = "log10",
            node_label_size_range = c(0.01, 0.02),
            # edge_size_trans = "log10",
            edge_size_range = c(0.008, 0.03) / 2,
            node_color = 2^abs(change) * sign(change),
            node_color_trans = "linear",
            node_color_range = diverging_palette(),
            node_color_interval = c(-4, 4),
            # edge_color_trans = "linear",
            # edge_color_range = diverging_palette(),
            # edge_color_interval =  c(-4, 4),
            # node_label_max = 500,
            node_color_axis_label = "Factor change",
            node_size_axis_label = "Number of genes",
            layout = "da", initial_layout = "re",
            output_file = result_path("gene_expression--cellular_component"))
```



### Biological Process

```{r cache = TRUE}
bp_res <- res[rep(1:nrow(res), sapply(bp_class, length)), ]
bp_res$class <- unlist(bp_class)
library(metacoder)

obj <- parse_tax_data(as.data.frame(bp_res), class_cols = "class", class_sep = "|")
obj$funcs <- c(obj$funcs,
                      change = function(x, subset = NULL) {
                        vapply(obs(x, "tax_data"),
                               function(i) {
                                 obs_change <- obj$data$tax_data[i, ]$log2FoldChange[obj$data$tax_data[i, ]$padj <= min_p_value]
                                 mean(obs_change, na.rm = TRUE)
                               },
                               numeric(1))
                      },
                      num_changed = function(x, subset = NULL) {
                        vapply(obs(x, "tax_data"),
                               function(i) {
                                 sum(obj$data$tax_data[i, ]$padj <= min_p_value, na.rm = TRUE)
                               },
                               numeric(1))
                      })
```

```{r cache = TRUE}
set.seed(7) #2, 4, 5, 7*, 19
mgsub <- function(pattern, replacement, x, ...) { # from: http://stackoverflow.com/questions/15253954/replace-multiple-arguments-with-gsub
  if (length(pattern)!=length(replacement)) {
    stop("pattern and replacement do not have the same length.")
  }
  result <- x
  for (i in 1:length(pattern)) {
    result <- gsub(pattern[i], replacement[i], result, ...)
  }
  result
}

to_replace <- matrix(ncol = 2, byrow = TRUE,
                     c("regulation of growth", "",
                       "activation of innate immune response", "",
                       "system development", "",
                       "regulation of response to stimulus", "",
                       "lipid metabolic process", "",
                       "selenium compound metabolic process", "selenium metabolic process"
                       ))
output_path <- file.path(output_folder,
                                    paste0("gene_expression--biological_process",
                                           output_format))
obj %>%
  filter_taxa(num_changed > 0) %>%
  filter_taxa(n_supertaxa <= 3) %>%
  mutate_obs("plot_data", 
             taxon_id = taxon_ids,
             plotted_name = gsub("_", " ", taxon_names),
              f_change = 2^abs(change) * sign(change)) %>%
  mutate_obs("plot_data",
             short_name = vapply(plotted_name, FUN.VALUE = character(1), function(x) {
               mgsub(pattern = to_replace[, 1], replacement =  to_replace[, 2], x, fixed = TRUE)
             })) %>%
  heat_tree(node_label = ifelse(abs(f_change) > 1, short_name, NA),
            node_size = num_changed,
            # node_size_trans = "log10",
            node_size_range = c(0.008, 0.03),
            # node_label_size_trans = "log10",
            node_label_size_range = c(0.012, 0.02),
            # edge_size_trans = "log10",
            edge_size_range = c(0.008, 0.03) / 2,
            node_color = f_change,
            node_color_trans = "linear",
            node_color_range = diverging_palette(),
            node_color_interval = c(-5, 5),
            edge_color_trans = "linear",
            edge_color_range = diverging_palette(),
            edge_color_interval =  c(-5, 5),
            node_color_axis_label = "Fold change",
            node_size_axis_label = "Number of genes",
            layout = "da", initial_layout = "re",
            output_file = result_path("gene_expression--biological_process"))
```


### Molecular Function

```{r cache = TRUE}
mf_res <- res[rep(1:nrow(res), sapply(mf_class, length)), ]
mf_res$class <- unlist(mf_class)
library(metacoder)
obj <- parse_tax_data(as.data.frame(mf_res), class_cols = "class", class_sep = "|")
obj$funcs <- c(obj$funcs,
                      change = function(x, subset = NULL) {
                        vapply(obs(x, "tax_data"),
                               function(i) {
                                 obs_change <- obj$data$tax_data[i, ]$log2FoldChange[obj$data$tax_data[i, ]$padj <= min_p_value]
                                 mean(obs_change, na.rm = TRUE)
                               },
                               numeric(1))
                      },
                      num_changed = function(x, subset = NULL) {
                        vapply(obs(x, "tax_data"),
                               function(i) {
                                 sum(obj$data$tax_data[i, ]$padj <= min_p_value, na.rm = TRUE)
                               },
                               numeric(1))
                      })
```


```{r cache = TRUE}
obj %>%
  filter_taxa(num_changed > 0) %>%
  filter_taxa(n_supertaxa <= 3) %>%
  # filter_taxa(n_supertaxa >= 1) %>% 
  # filter_taxa(nchar(taxon_names) <= 40) %>%
  heat_tree(node_label = ifelse(nchar(taxon_names) <= 50, taxon_names, ""),
            node_size = num_changed,
            # node_size_trans = "log10",
            node_size_range = c(0.01, 0.03),
            # node_label_size_trans = "log10",
            node_label_size_range = c(0.01, 0.015),
            # edge_size_trans = "log10",
            edge_size_range = c(0.008, 0.03) / 2,
            node_color = 2^abs(change) * sign(change),
            node_color_trans = "linear",
            node_color_range = diverging_palette(),
            node_color_interval = c(-4, 4),
            edge_color_trans = "linear",
            edge_color_range = diverging_palette(),
            edge_color_interval =  c(-4, 4),
            node_color_axis_label = "Factor change",
            node_size_axis_label = "Number of genes",
            layout = "da", initial_layout = "re",
            output_file = result_path("gene_expression--molecular_function"))
```


## Software and packages used

```{r}
sessionInfo()
```
