---
title: "Tara oceans example"
bibliography: bibliography.bibtex
---

```{r init, echo = FALSE, message = FALSE}
library(knitr)
if (! is.null(current_input())) { # if knitr is being used
  knitr::read_chunk("settings.R")
} else {
  source("settings.R")
}
```

```{r rendering_settings, include=FALSE}
```

## Requirements 

**NOTE:** This analysis requires at least 10GB of RAM to run.
It uses large files not included in the repository and many steps can take a few minutes to run. 

## Parameters

```{r io_settings}
```

## Introduction

Here we present the code to reproduce the TARA Oceans figure in the metacoder publication. 
The TARA oceans expedition sampled ocean water from around the world and conducted 18S metabarcoding and metagenomics to infer species composition [@de2015eukaryotic]. 
Their paper mentioned that a large portion of the diversity found could not be well identified with currently available reference sequences. 
We decided to look at this using heat trees to see which taxonomic groups are poorly represented in reference databases. 

## Parsing taxonomic data

The data set at the below URL was downloaded and uncompressed:

http://taraoceans.sb-roscoff.fr/EukDiv/data/Database_W5_OTU_occurences.tsv.zip

```{r}
library(metacoder)
```


```{r tara_parse, warning=FALSE, message=FALSE, cache = TRUE}
raw_data <- readr::read_tsv(file.path(input_folder, "Database_W5_OTU_occurences.tsv"))
obj <- parse_tax_data(raw_data, class_cols = "lineage", class_sep = "\\|", sep_is_regex = TRUE)
```


## Getting sample data

The sample data was downloaded from the URL below:

http://taraoceans.sb-roscoff.fr/EukDiv/data/Database_W1_Sample_parameters.xls

```{r}
library(readxl)
```

```{r tara_parse_sample, cache = TRUE}
sample_data <- read_excel(file.path(input_folder, "Database_W1_Sample_parameters.xls"))
```


## Caluculate read abundance per taxon

The input data included read abundance for each sample-OTU combination, but we need the abundances associated with each taxon for graphing.
The `obs` function is used here to get a list of row indexes corresponding to OTUs for each taxon.
There will usually be multiple OTUs assigned to the same taxon, especially for coarse taxonomic ranks (e.g. the root will have all OTU indexes), so the abundances at those indexes are are summed to provide the total abundance for each taxon.

```{r tara_calculate_abundance, cache = TRUE}
obj$data$tax_abund <- calc_taxon_abund(obj, data = "tax_data",
                                       cols = sample_data[["PANGAEA ACCESSION NUMBER"]])
```


## Plot read and OTU abundance

First lets get an idea of the taxonomic distribution of reads and OTUs. 
This data set has many taxa (`r length(taxon_names(obj))`) and deep taxonomies (a max depth of `r max(n_supertaxa(obj) + 1)`), so graphing all of it effectively will be a challenge. 
To make it more likly that the layout algorithms can untangle it all, we filter out taxa with less than 100 reads and remove some of the coarse taxonomic ranks so that we get multiple trees per graph.
Each tree is then easier to layout for the computer and easier to read for us humans (assuming you saved it as a PDF and zoom in).
Additionally, we will remove some taxa that have ambiguous names since they do not provide any information. 
Note that we are not throwing out read or OTU counts when we do this filtering; the reads and OTU counts still apply to supertaxa (e.g. Metazoa) that were not filtered out. 

```{r tara_plot_all, cache = TRUE}
seed = 9 #9, 10, 12 is good
set.seed(seed)
taxa_patterns_to_hide <- paste0("^", c("[X]+", "X\\+sp\\.", "NA", "root", "\\*", "sp\\.", "sp"), "$")
taxa_patterns_to_remove <- paste0("^", c("X\\+sp\\.", "NA", "root", "\\*", "sp\\.", "sp"), "$")
background_color <- "#ccfff7"
obj %>%
  mutate_obs("tax_abund", abundance = rowSums(obj$data$tax_abund[sample_data[["PANGAEA ACCESSION NUMBER"]]])) %>%
  filter_taxa(taxon_names == "Eukaryota", supertaxa = TRUE, invert = TRUE) %>%
  filter_taxa(abundance >= 100) %>%
  filter_taxa(! Reduce(`|`, lapply(taxa_patterns_to_remove, grepl, x = taxon_names))) %>%
  heat_tree(title = "Plankton diversity in the sunlit ocean",
            title_size = 0.03,
            node_color_axis_label = "Number of reads (Abundance)",
            node_size_axis_label = "Number of species (OTUs)",
            node_size = n_obs,
            node_size_range = c(0.0012, NA),
            node_color = abundance,
            node_color_range = c("grey", "#80cdc1", "#018571", "#dfc27d", "#a6611a"),
            node_color_trans = "log10",
            node_label = ifelse(grepl(pattern = "^[a-zA-z\\-]{1,25}$", taxon_names) &
                                  ! Reduce(`|`, lapply(taxa_patterns_to_hide, grepl, x = taxon_names)) &
                                  abundance > 10000,
                                taxon_names, ""),
            node_label_color = "#000000",
            node_label_color_trans = "area",
            node_label_size = (n_obs / (n_supertaxa + 1)) ^ 0.5,
            node_label_size_trans = "area",
            node_label_size_range = c(0.001, NA),
            tree_label = taxon_names,
            tree_label_color = "#00806c",
            tree_label_max = 100,
            initial_layout = "re", layout = "da",
            overlap_avoidance = .65,
            background_color = background_color,
            # maxiter = 50, fineiter = 50,
            output_file = result_path("sup_figure_1--tara_all_plankton_diversity"))
```


## Plot propotion of OTUs identified 

Next we will look at how well the reference database used was able to classify OTUs. 
This will give us an idea of how well current references databases account for the diversity of eukaryotes in the oceans. 
For each taxon we will identify the proportion of OTUs with less than a 90% match to the most similar reference sequence to approximate classification certainty. 

```{r tara_calc_identified, cache = TRUE}
# color_range <- c("firebrick1", "orangered",  "orange", "yellow3", "yellowgreen", "green")
color_range <- c(viridis::plasma(10)[4:9], "lightgrey")
obj <- mutate_obs(obj, data = "tax_abund", mean_pid = vapply(obs(obj, data = "tax_data"),
                                                 function(x) mean(obj$data$tax_data$pid[x], na.rm = TRUE), numeric(1)))
# Percentage of OTUs with less than 90% idententiy
obj <- mutate_obs(obj, "tax_abund", percent_known = vapply(obs(obj, data = "tax_data"),
                                                 function(x) sum(obj$data$tax_data$pid[x] >= 90, na.rm = TRUE) / length(x) * 100, numeric(1)))
```

We can plot this the same way that we plotted OTU and read abundance.
In fact, we can preserve abundance information by letting the size of nodes and edges correspond to OTU and read abundance and using color to plot classification certainty. 
Since this figure will be part of the publication, we have not added node labels since they would not be visible at the scale publication figures are viewed.


```{r tara_plot_all_identified, cache = TRUE}
seed = 13
set.seed(seed)
taxa_patterns_to_hide <- paste0("^", c("[X]+", "X\\+sp\\.", "NA", "root", "\\*", "sp\\.", "sp"), "$")
taxa_patterns_to_remove <- paste0("^", c("[X]+", "X\\+sp\\.", "NA", "root", "\\*", "sp\\.", "sp"), "$")
identified_plot <- obj %>%
  mutate_obs("tax_abund", abundance = rowSums(obj$data$tax_abund[sample_data[["PANGAEA ACCESSION NUMBER"]]])) %>%
  filter_taxa(taxon_names == "Eukaryota", supertaxa = TRUE, invert = TRUE) %>%
  filter_taxa(abundance >= 100) %>%
  filter_taxa(! Reduce(`|`, lapply(taxa_patterns_to_remove, grepl, x = taxon_names))) %>%
  filter_taxa(n_subtaxa < 30 & n_supertaxa == 0, subtaxa = TRUE, invert = TRUE) %>%
  heat_tree(
    title_size = 0.03,
    node_color_axis_label = "Percent of species (OTUs) identified",
    node_size_axis_label = "Number of species (OTUs)",
    node_size = n_obs,
    node_size_range = c(0.0012, NA),
    node_color = percent_known,
    node_color_range = color_range,
    node_color_trans = "linear",
    node_color_interval = c(0, 100),
    node_label_size = (n_obs / (n_supertaxa + 1)) ^ 0.5,
    node_label_size_trans = "area",
    node_label_size_range = c(0.001, NA),
    tree_label = taxon_names,
    tree_label_size_range = c(0.008, 0.035),
    initial_layout = "re", layout = "da",
    overlap_avoidance = .65,
    # maxiter = 50, fineiter = 50,
    output_file = result_path("sup_figure_2--tara_proportion_identified"))
identified_plot
```


### Just Metazoa

We can also plot a single tree for a little less overwhelming graph.
Below, we plot only the metazoa using the same meanings for color and size.
This is also a publication figure, so taxa have been filtered out more aggressively and only a few labels are shown. 

```{r tara_plot_metazoa_identified, cache = TRUE}
seed <- 21
set.seed(seed)
taxa_patterns_to_hide <- paste0("^", c("[X]+", "X\\+sp\\.", "NA", "root", "\\*", "sp\\.", "sp", "neglecta", "Bothriocroton", "Thysanoessa", "Metridia", "Aidanosagitta", "Cephalopoda", "Craniata"), "$")
taxa_patterns_to_remove <- paste0("^", c("[X]+", "X\\+sp\\.", "NA", "root", "\\*", "sp\\.", "sp"), "$")
metazoa_plot <- obj %>%
  mutate_obs("tax_abund", abundance = rowSums(obj$data$tax_abund[sample_data[["PANGAEA ACCESSION NUMBER"]]])) %>%
  filter_taxa(abundance >= 3100) %>%
  filter_taxa(! Reduce(`|`, lapply(taxa_patterns_to_remove, grepl, x = taxon_names))) %>%
  filter_taxa(taxon_names == "Metazoa", subtaxa = TRUE) %>%
  heat_tree(node_color_axis_label = "Percent of OTUs identified",
            edge_color_axis_label = "Percent of OTUs identified",
            edge_size = abundance,
            edge_size_range = c(0.0005, 0.013),
            edge_size_trans = "area",
            edge_color = percent_known,
            edge_color_range = color_range,
            edge_color_trans = "linear",
            node_size_axis_label = "Number of OTUs",
            edge_size_axis_label = "Number of reads",
            node_size = n_obs,
            node_size_range = c(0.005, 0.03),
            node_color = percent_known,
            node_color_range = color_range,
            node_color_trans = "linear",
            node_color_interval = c(0, 100),
            node_label = ifelse(grepl(pattern = "^[a-zA-z\\-]{1,25}$", taxon_names) &
                                  ! Reduce(`|`, lapply(taxa_patterns_to_hide, grepl, x = taxon_names)),
                                taxon_names, ""),
            node_label_size = n_obs,
            node_label_size_range = c(0.013, 0.018),
            node_label_size_trans = "area",
            node_label_max = 100,
            initial_layout = "fr", layout = "da",
            output_file = result_path(paste0("figure_1--tara_metazoa--seed_", seed)))
print(metazoa_plot)
```


### Combine plots

Finally, lets combine the two graphs to produce the publication figure.

```{r}
library(gridExtra)
library(cowplot)
```

```{r, fig.width=7, fig.height=10, cache = TRUE}
combined_plot <- plot_grid(metazoa_plot, identified_plot, ncol = 1, nrow = 2, labels = c("a", "b"))
save_plot(filename = result_path("figure_1--tara_metazoa_combined"),
          combined_plot, ncol = 1, nrow = 2, base_height = 5, base_width = 8)
combined_plot
```

## Software and packages used

```{r, cache = TRUE}
sessionInfo()
```

## References