<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title></title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-84254038-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><div class="logo_frame"><span class="logo_helper"></span><img src="images/hex_sticker.png" height=50px /></div> metacoder</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="example.html">Example</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="workshop--00--introduction.html">Introduction</a>
    </li>
    <li>
      <a href="workshop--00--required_software.html">Required software</a>
    </li>
    <li>
      <a href="workshop--00--required_data.html">Required datasets</a>
    </li>
    <li>
      <a href="workshop--03--parsing.html">Getting data into R</a>
    </li>
    <li>
      <a href="workshop--04--manipulating.html">Manipulating taxonomic data</a>
    </li>
    <li>
      <a href="workshop--05--plotting.html">Plotting taxonomic data</a>
    </li>
    <li>
      <a href="workshop--06--quality_control.html">Data quality control</a>
    </li>
    <li>
      <a href="workshop--07--diversity_stats.html">Diversity and Ordination</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Publication
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404">PLOS Comp Bio paper</a>
    </li>
    <li>
      <a href="publication--05--16s_human_microbiome.html">Human microbiome</a>
    </li>
    <li>
      <a href="publication--06--tara_oceans.html">TARA Oceans</a>
    </li>
    <li>
      <a href="publication--01--silva.html">Digital PCR: SILVA</a>
    </li>
    <li>
      <a href="publication--02--rdp.html">Digital PCR: RDP</a>
    </li>
    <li>
      <a href="publication--03--greengenes.html">Digital PCR: Greengenes</a>
    </li>
    <li>
      <a href="publication--04--16s_database_comparision.html">Digital PCR: Comparing databases</a>
    </li>
    <li>
      <a href="publication--08--voting.html">Voting Geography</a>
    </li>
    <li>
      <a href="publication--09--gene_expression.html">Gene Expression</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Appendices
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="appendix--00--intro_to_r.html">Introduction to R</a>
    </li>
    <li>
      <a href="appendix--00--intro_to_rstudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="appendix--00--glossary.html">Glossary</a>
    </li>
    <li>
      <a href="appendix--00--presentation_page.html">Slide show</a>
    </li>
    <li>
      <a href="poster/poster.pdf">Poster</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder">Metacoder source code</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder_documentation">Website source code</a>
    </li>
  </ul>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<div id="data-quality-control" class="section level1">
<h1>Data quality control</h1>
<p>Sequencing technologies all have some amount of error. Some of this error happens during PCR and some happens during sequencing. There are lots of ways to filter out errors including:</p>
<ul>
<li>Removing sequences/bases that the sequencer indicates are low quality. This information is typically in the <a href ="appendix--00--glossary.html#fastq_anchor">.fastq</a> files returned by the sequencer. There are many programs, such as <a href="http://www.usadellab.org/cms/?page=trimmomatic">trimmomatic</a>, that can use the quality scores in these files to filter out low-quality sequences.</li>
<li>Clustering similar sequences together. Many erroneous sequences are very similar to the true sequences, so they will be incorporated into the same <a href ="appendix--00--glossary.html#operational_taxonomic_units_(otus)_anchor">OTU</a> as the true sequence. In that sense, clustering into OTUs will hide some sequencing error.</li>
<li>Removal of <a href ="appendix--00--glossary.html#chimeric_sequences_anchor">chimeric sequences</a>. “Chimeras” are more significant errors that occur during PCR when an incomplete amplicon acts as a primer for a different template in a subsequent cycle. Most amplicon metagenomic pipelines (e.g. QIIME, mothur, usearch) have functions to detect and remove chimeras.</li>
<li>Removing low-abundance sequences. Most of the time, erroneous sequences will occur much less often than the true sequence. Simply removing any unique sequence / OTU that appears less than some minimum number of times is an effective way to remove these errors. OTUs represented by only a single sequence are commonly called <a href ="appendix--00--glossary.html#singleton_anchor">singletons</a>. So when you hear things like “singletons were removed” and means all OTUs with only a single sequence were removed.</li>
<li><a href ="appendix--00--glossary.html#rarefaction_anchor">Rarefying</a> read counts. Usually, we try to make each sample in a sequencing run have the same number of reads sequenced, but that does not always happen. When a sample has many more reads than another sample, its apparent diversity can be artificially inflated since rare taxa are more likely to be found. To avoid this, the reads are subsampled to a fixed number or “rarefied”.</li>
<li>Converting counts to presence/absence. Due to PCR biases and the nature of <a href ="appendix--00--glossary.html#compositional_data_anchor">compositional data</a>, many researchers advocate not using read count information at all. Instead, they recommend converting counts to simply “present” or “absent”.</li>
</ul>
<p>Since we are only working with abundance matrices in these tutorials, the first three types of quality control have already been done (ideally), but we can still remove low-abundance OTUs and rarefy the counts of the remaining OTUs.</p>
<div id="load-example-data" class="section level2">
<h2>Load example data</h2>
<p>If you are starting the workshop at this section, or had problems running code in a previous section, use the following to load the data used in this section. You can download the “filtered_data.Rdata” file <a href="filtered_data.Rdata" download="filtered_data.Rdata">here</a>. If <code>obj</code> and <code>sample_data</code> are already in your environment, you can ignore this and proceed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;filtered_data.Rdata&quot;</span>)</code></pre></div>
</div>
<div id="removing-low-abundance-counts" class="section level2">
<h2>Removing low-abundance counts</h2>
<p>The easiest way to get rid of some error in your data is to throw out any count information below some threshold. The threshold is up to you; removing singletons or <a href ="appendix--00--glossary.html#doubletons_anchor">doubletons</a> is common, but lets be more conservative and remove any counts less than 10. The <code>zero_low_counts</code> counts can be used to convert any counts less than some number to zero.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(metacoder)
obj<span class="op">$</span>data<span class="op">$</span>otu_counts &lt;-<span class="st"> </span><span class="kw">zero_low_counts</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="dt">min_count =</span> <span class="dv">10</span>,
                                       <span class="dt">other_cols =</span> <span class="ot">TRUE</span>) <span class="co"># keep OTU_ID column</span></code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Zeroing 576512 of 6331728 counts less than 10.</code></pre>
<p>That set all read counts less than 10 to zero. It did not filter out any OTUs or their associated taxa however. We can do that using <code>filter_obs</code>, which is used to filter data associated with a taxonomy in a <code>taxmap</code> object. First lets find which OTUs now have now reads associated with them.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">no_reads &lt;-<span class="st"> </span><span class="kw">rowSums</span>(obj<span class="op">$</span>data<span class="op">$</span>otu_counts[, sample_data<span class="op">$</span>SampleID]) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>
<span class="kw">sum</span>(no_reads) <span class="co"># when `sum` is used on a TRUE/FALSE vector it counts TRUEs</span></code></pre></div>
<pre class="r-output"><code>## [1] 14990
</code></pre>
<p>So now 14990 of 21684 OTUs have no reads. We can remove them and the taxa they are associated with like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj &lt;-<span class="st"> </span><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="op">!</span><span class="st"> </span>no_reads, <span class="dt">drop_taxa =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   489 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>ccz</span><span>. Enterococcaceae, </span><span style='color: #00BB00;'>cfp</span><span>. Dermacoccaceae
##   489 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>byk</span><span>, </span><span style='color: #00BB00;'>aqy</span><span>-&gt;</span><span style='color: #00BB00;'>ccz</span><span>, </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cfp</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 6,694 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2              0          0         0         0         0         0
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3          10251         59     15857       309       134        90
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4           1447       2496      2421      3331       143      1847
##       # ... with 6,691 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<div id="rarefaction" class="section level2">
<h2>Rarefaction</h2>
<p>Rarefaction is used to simulate even numbers of samples (i.e. reads). Even sampling is important for at least two reasons:</p>
<ul>
<li>When comparing diversity of samples, more samples make it more likely to observe rare species. This will have a larger effect on some diversity indexes than others, depending on how they weigh rare species.</li>
<li>When comparing the similarity of samples, the presence of rare species due to higher sampling depth in one sample but not another can make the two samples appear more different than they actually are.</li>
</ul>
<p>Therefore, when comparing the diversity or similarity of samples, it is important to rarefy, or subsample, to a constant depth. Typically, the depth chosen is the minimum sample depth. If the minimum depth is very small, the samples with the smallest depth can be removed and the minimum depth of the remaining samples can be used.</p>
<p>Lets take a look at the distribution of read depths of our samples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(<span class="kw">colSums</span>(obj<span class="op">$</span>data<span class="op">$</span>otu_counts[, sample_data<span class="op">$</span>SampleID]))</code></pre></div>
<p><img src="workshop--06--quality_control_files/figure-html/unnamed-chunk-5-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>We have a minimum depth of 1857, a median of 78767 and a maximum depth of 1.8935310^{5}. We could try to remove one or two of the samples with the smallest depth, since it seems like a waste to throw out so much data, but for this tutorial we will just rarefy to the minimum of 1857 reads. The <code>vegan</code> package implements many functions to help with rarefying data, but we will use a function from <code>metacoder</code> that calls the functions from <code>vegan</code> and reformats the input and outputs to make them easier to use with the way our data is formatted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>otu_rarefied &lt;-<span class="st"> </span><span class="kw">rarefy_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="dt">other_cols =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Rarefying to 1857 since that is the lowest sample total.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   489 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>ccz</span><span>. Enterococcaceae, </span><span style='color: #00BB00;'>cfp</span><span>. Dermacoccaceae
##   489 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>byk</span><span>, </span><span style='color: #00BB00;'>aqy</span><span>-&gt;</span><span style='color: #00BB00;'>ccz</span><span>, </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cfp</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 6,694 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2              0          0         0         0         0         0
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3          10251         59     15857       309       134        90
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4           1447       2496      2421      3331       143      1847
##       # ... with 6,691 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>otu_rarefied</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 6,694 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2              0          0         0         0         0         0
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3            398          1       402         6       134         3
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4             52         35        70        66       143        30
##       # ... with 6,691 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
<p>This probably means that some OTUs now have no reads in the rarefied dataset. Lets remove those like we did before. However, since there is now two tables, we should not remove any taxa since they still might be used in the rarefied table, so we will not add the <code>drop_taxa = TRUE</code> option this time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">no_reads &lt;-<span class="st"> </span><span class="kw">rowSums</span>(obj<span class="op">$</span>data<span class="op">$</span>otu_rarefied[, sample_data<span class="op">$</span>SampleID]) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>
obj &lt;-<span class="st"> </span><span class="kw">filter_obs</span>(obj, <span class="st">&quot;otu_rarefied&quot;</span>, <span class="op">!</span><span class="st"> </span>no_reads)
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   489 taxa: <span style='color: #00BB00;'>aad</span><span>. Bacteria, </span><span style='color: #00BB00;'>aaf</span><span>. Proteobacteria ... </span><span style='color: #00BB00;'>ccz</span><span>. Enterococcaceae, </span><span style='color: #00BB00;'>cfp</span><span>. Dermacoccaceae
##   489 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aaf</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aag</span><span>, </span><span style='color: #00BB00;'>aad</span><span>-&gt;</span><span style='color: #00BB00;'>aah</span><span> ... </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>byk</span><span>, </span><span style='color: #00BB00;'>aqy</span><span>-&gt;</span><span style='color: #00BB00;'>ccz</span><span>, </span><span style='color: #00BB00;'>amj</span><span>-&gt;</span><span style='color: #00BB00;'>cfp</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 6,694 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2              0          0         0         0         0         0
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3          10251         59     15857       309       134        90
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4           1447       2496      2421      3331       143      1847
##       # ... with 6,691 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##     </span><span style='font-weight: bold;'>otu_rarefied</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 5,053 x 294</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>ben</span><span>      2              0          0         0         0         0         0
##       2 </span><span style='color: #00BB00;'>beo</span><span>      3            398          1       402         6       134         3
##       3 </span><span style='color: #00BB00;'>ben</span><span>      4             52         35        70        66       143        30
##       # ... with 5,050 more rows, and 286 more variables: M1981P587 &lt;dbl&gt;,
##       #   M1982P729 &lt;dbl&gt;, M1979P447 &lt;dbl&gt;, M1982P701 &lt;dbl&gt;, M1982P661 &lt;dbl&gt;,
##       #   M1978P293 &lt;dbl&gt;, M1980P462 &lt;dbl&gt;, M1555P221 &lt;dbl&gt;, M1980P471 &lt;dbl&gt;,
##       #   M1979P422 &lt;dbl&gt;, …
##   0 functions:
</span></code></pre>
</div>
<div id="rarefaction-curves" class="section level2">
<h2>Rarefaction curves</h2>
<p>The relationship between the number of reads and the number of OTUs can be described using <a href ="appendix--00--glossary.html#rarefaction_anchor">rarefaction curves</a>. Each line represents a different sample (i.e. column) and shows how many OTUs are found in a random subsets of different numbers of reads. The <code>rarecurve</code> function from the <code>vegan</code> package will do the random subsampling and plot the results for an abundance matrix. For example, the code below plots the rarefaction curve for a single sample:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(vegan)</code></pre></div>
<pre><code>## Loading required package: permute</code></pre>
<pre><code>## Loading required package: lattice</code></pre>
<pre><code>## This is vegan 2.5-2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rarecurve</span>(<span class="kw">t</span>(obj<span class="op">$</span>data<span class="op">$</span>otu_counts[, <span class="st">&quot;M1981P563&quot;</span>]), <span class="dt">step =</span> <span class="dv">20</span>,
          <span class="dt">sample =</span> <span class="kw">min</span>(<span class="kw">colSums</span>(obj<span class="op">$</span>data<span class="op">$</span>otu_counts[, sample_data<span class="op">$</span>SampleID])),
          <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>, <span class="dt">cex =</span> <span class="fl">1.5</span>)</code></pre></div>
<p><img src="workshop--06--quality_control_files/figure-html/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>For this sample, few new OTUs are observed after about 20,000 reads. Since the number of OTUs found flattens out, it suggests that the sample had sufficient reads to capture most of the diversity. Like other functions in <code>vegan</code>, <code>rarecurve</code> expects samples to in rows instead of columns, so we had to <a href ="appendix--00--glossary.html#transpose_anchor">transpose</a> (make rows columns) the matrix with the <code>t</code> function before passing it to <code>rarecurve</code>. If you plot all the samples, which takes a few minutes, it looks like this:</p>
<p><img src="workshop--06--quality_control_files/figure-html/unnamed-chunk-9-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="converting-to-presenceabsence" class="section level2">
<h2>Converting to presence/absence</h2>
<p>Many researchers believe that read depth is not informative due to PCR and sequencing biases. Therefore, instead of comparing read counts, the counts can be converted to presence or absence of an OTU in a given sample. This can be done like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">counts_to_presence</span>(obj, <span class="st">&quot;otu_rarefied&quot;</span>)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 5,053 x 293</span><span>
##    taxon_id M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611 M1981P587 M1982P729
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span>    
## </span><span style='color: #BCBCBC;'> 1</span><span> ben      FALSE     FALSE      FALSE     FALSE     FALSE     FALSE     TRUE      FALSE    
## </span><span style='color: #BCBCBC;'> 2</span><span> beo      TRUE      TRUE       TRUE      TRUE      TRUE      TRUE      TRUE      TRUE     
## </span><span style='color: #BCBCBC;'> 3</span><span> ben      TRUE      TRUE       TRUE      TRUE      TRUE      TRUE      TRUE      TRUE     
## </span><span style='color: #BCBCBC;'> 4</span><span> bep      TRUE      TRUE       TRUE      TRUE      FALSE     TRUE      TRUE      TRUE     
## </span><span style='color: #BCBCBC;'> 5</span><span> beq      TRUE      TRUE       TRUE      TRUE      FALSE     TRUE      TRUE      TRUE     
## </span><span style='color: #BCBCBC;'> 6</span><span> amk      TRUE      TRUE       TRUE      TRUE      TRUE      TRUE      TRUE      TRUE     
## </span><span style='color: #BCBCBC;'> 7</span><span> bes      TRUE      TRUE       FALSE     TRUE      FALSE     TRUE      TRUE      TRUE     
## </span><span style='color: #BCBCBC;'> 8</span><span> bet      TRUE      TRUE       TRUE      TRUE      TRUE      FALSE     TRUE      FALSE    
## </span><span style='color: #BCBCBC;'> 9</span><span> beu      TRUE      FALSE      TRUE      TRUE      FALSE     TRUE      TRUE      FALSE    
## </span><span style='color: #BCBCBC;'>10</span><span> bev      TRUE      TRUE       FALSE     TRUE      FALSE     TRUE      TRUE      TRUE     
## </span><span style='color: #949494;'># ... with 5,043 more rows, and 284 more variables: M1979P447 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1982P701 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>,
## #   M1982P661 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1978P293 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1980P462 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1555P221 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1980P471 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>,
## #   M1979P422 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1982P655 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, M1554P182 </span><span style='color: #949494;font-style: italic;'>&lt;lgl&gt;</span><span style='color: #949494;'>, …</span><span>
</span></code></pre>
<p>For this workshop however, we will use the read counts.</p>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<p>In these exercises, we will be using the <code>obj</code> from the analysis above. If you did not run the code above or had problems, run the following code to get the objects used. You can download the “clean_data.Rdata” file <a href="clean_data.Rdata" download="clean_data.Rdata">here</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">&quot;clean_data.Rdata&quot;</span>)</code></pre></div>
<p><strong>1a)</strong> Consider the following rarefaction curves:</p>
<p><img src="workshop--06--quality_control_files/figure-html/unnamed-chunk-13-1.png" width="960" style="display: block; margin: auto;" /></p>
<p><strong>1b)</strong> Which sample is more diverse?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-14">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-14" class="collapse">
B
</div>
<p><br /></p>
<p><strong>1c)</strong> Which sample needs the most reads to capture all of its diversity?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-15">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-15" class="collapse">
B
</div>
<p><br /></p>
<p><strong>1d)</strong> What would be a sufficient read count to capture all of the diversity in these three samples?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-16">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-16" class="collapse">
Around 40,000 reads.
</div>
<p><br /></p>
<p><strong>2)</strong> Look at the documentation for <code>rarefy_obs</code>. Rarefy the sample OTU counts to 1000 reads.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-17">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-17" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rarefy_obs</span>(obj, <span class="st">&quot;otu_counts&quot;</span>, <span class="dt">sample_size =</span> <span class="dv">1000</span>)</code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 6,694 x 293</span><span>
##    taxon_id M1981P563 M1977P1709 M1980P502 M1981P606 M1981P599 M1981P611 M1981P587 M1982P729
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>        </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> ben              0          0         0         0         0         0         1         0
## </span><span style='color: #BCBCBC;'> 2</span><span> beo            170          0       252         4        76         1       191         2
## </span><span style='color: #BCBCBC;'> 3</span><span> ben             28         16        40        25        77        16        32        16
## </span><span style='color: #BCBCBC;'> 4</span><span> bep              2          0         5        10         0         2         5         3
## </span><span style='color: #BCBCBC;'> 5</span><span> beq              2         27         3        28         0        39         3        22
## </span><span style='color: #BCBCBC;'> 6</span><span> amk             42          3        76        10       122         3        40         3
## </span><span style='color: #BCBCBC;'> 7</span><span> bes              0         29         0        11         0        26         4        17
## </span><span style='color: #BCBCBC;'> 8</span><span> bet              1          1         2         1         9         1        10         0
## </span><span style='color: #BCBCBC;'> 9</span><span> beu             82          0        74         3         0         1        64         0
## </span><span style='color: #BCBCBC;'>10</span><span> bev              0         10         0        20         0        16         0        31
## </span><span style='color: #949494;'># ... with 6,684 more rows, and 284 more variables: M1979P447 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1982P701 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1982P661 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1978P293 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1980P462 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1555P221 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1980P471 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1979P422 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1982P655 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1554P182 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, …</span><span>
</span></code></pre>
</div>
<p><br /></p>
<p><strong>3)</strong> What are two ways of accounting for uneven sample depth?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-18">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-18" class="collapse">
Rarefaction and converting counts to proportions.
</div>
<p><br /></p>
<p><strong>4)</strong> What are some reasons that read counts might not correspond to species abundance in the original sample? Try to think of at least 3. There are 7 listed in the answer below and there are probably other reasons not listed.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-19">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-19" class="collapse">
<ul>
<li>PCR primers amplify some species more than others.</li>
<li>DNA sequencers sequence some sequences more than others.</li>
<li>The locus being amplified might have different numbers of copies in different organisms.</li>
<li>DNA from dead organisms can still be amplified, although this might not be a problem depending on the goal of the research.</li>
<li>Different sequences/organisms degrade in the environment at different rates.</li>
<li>Some organisms (e.g. spore-forming bacteria) are more resistant to DNA extraction techniques, so might not be as well represented in the DNA extract.</li>
<li>Some species have more DNA per individual / weight than others.
</div>
<br /></li>
</ul>
</div>
</div>

<div class="footer_section">
  <div class="footer_content">
    <span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type">Metacoder's documentation</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://grunwaldlab.cgrb.oregonstate.edu/" property="cc:attributionName" rel="cc:attributionURL">The Grunwald lab and the USDA Horticultural Crops Research Unit</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/grunwaldlab/metacoder_documentation" rel="dct:source">https://github.com/grunwaldlab/metacoder_documentation</a>.
  </div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
