<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>appendix--00--no_internet_install.utf8.md</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-84254038-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><div class="logo_frame"><span class="logo_helper"></span><img src="images/hex_sticker.png" height=50px /></div> metacoder</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="example.html">Example</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="workshop--00--introduction.html">Introduction</a>
    </li>
    <li>
      <a href="workshop--00--required_software.html">Required software</a>
    </li>
    <li>
      <a href="workshop--00--required_data.html">Required datasets</a>
    </li>
    <li>
      <a href="workshop--03--parsing.html">Getting data into R</a>
    </li>
    <li>
      <a href="workshop--04--manipulating.html">Manipulating taxonomic data</a>
    </li>
    <li>
      <a href="workshop--05--plotting.html">Plotting taxonomic data</a>
    </li>
    <li>
      <a href="workshop--06--quality_control.html">Data quality control</a>
    </li>
    <li>
      <a href="workshop--07--diversity_stats.html">Diversity and Ordination</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Publication
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404">PLOS Comp Bio paper</a>
    </li>
    <li>
      <a href="publication--05--16s_human_microbiome.html">Human microbiome</a>
    </li>
    <li>
      <a href="publication--06--tara_oceans.html">TARA Oceans</a>
    </li>
    <li>
      <a href="publication--01--silva.html">Digital PCR: SILVA</a>
    </li>
    <li>
      <a href="publication--02--rdp.html">Digital PCR: RDP</a>
    </li>
    <li>
      <a href="publication--03--greengenes.html">Digital PCR: Greengenes</a>
    </li>
    <li>
      <a href="publication--04--16s_database_comparision.html">Digital PCR: Comparing databases</a>
    </li>
    <li>
      <a href="publication--08--voting.html">Voting Geography</a>
    </li>
    <li>
      <a href="publication--09--gene_expression.html">Gene Expression</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Appendices
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="appendix--00--intro_to_r.html">Introduction to R</a>
    </li>
    <li>
      <a href="appendix--00--intro_to_rstudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="appendix--00--glossary.html">Glossary</a>
    </li>
    <li>
      <a href="appendix--00--presentation_page.html">Slide show</a>
    </li>
    <li>
      <a href="appendix--00--poster_page.html">Poster</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder">Metacoder source code</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder_documentation">Website source code</a>
    </li>
  </ul>
</li>
<li>
  <a href="faq.html">FAQ</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<p>This page contains instructions on how to install the dependencies for this workshop without internet access. If you are doing this, you are probably reading this page from a file provided by the instructor rather than over the internet. Along with the website files (what you are reading now), there should be a directory called “software” that contains the software to install. <strong>If you have internet access, use the standard way of installing software described <a href="workshop--00--required_software.html">here</a></strong></p>
<div id="installing-r" class="section level2">
<h2>Installing R</h2>
<p>For Mac and Windows, use the installer included in the “software/r/mac” or “software/r/windows”. This tutorial does currently not provide a way to install R on Linux offline.</p>
</div>
<div id="installing-rstudio" class="section level2">
<h2>Installing Rstudio</h2>
<p>Use the installers in “software/rstudio/mac,”software/rstudio/windows&quot; or “software/rstudio/linux”, depending on your operating system.</p>
</div>
<div id="installing-r-packages" class="section level2">
<h2>Installing R packages</h2>
<p>The following packages are used in this workshop:</p>
<pre class="r-output"><code> [1] "vegan"     "metacoder" "taxa"      "ggplot2"   "dplyr"     "readr"     "stringr"   "agricolae"
 [9] "ape"       "phyloseq" 
</code></pre>
<p>However, these depend on many other packages that must also be installed, so the complete list is:</p>
<pre class="r-output"><code>  [1] "vegan"        "metacoder"    "taxa"         "ggplot2"      "dplyr"        "readr"       
  [7] "stringr"      "agricolae"    "ape"          "phyloseq"     "permute"      "lattice"     
 [13] "MASS"         "cluster"      "mgcv"         "graphics"     "grDevices"    "stats"       
 [19] "utils"        "grid"         "methods"      "nlme"         "Matrix"       "splines"     
 [25] "igraph"       "scales"       "taxize"       "seqinr"       "reshape2"     "zoo"         
 [31] "traits"       "RColorBrewer" "RCurl"        "reshape"      "lazyeval"     "magrittr"    
 [37] "rlang"        "biomformat"   "phylotate"    "ggfittext"    "ggrepel"      "cowplot"     
 [43] "GA"           "Rcpp"         "crayon"       "svglite"      "viridisLite"  "tibble"      
 [49] "tools"        "parallel"     "plyr"         "jsonlite"     "rhdf5"        "gtable"      
 [55] "assertthat"   "glue"         "pkgconfig"    "R6"           "tidyselect"   "foreach"     
 [61] "iterators"    "cli"          "stringi"      "shades"       "digest"       "withr"       
 [67] "bitops"       "hms"          "clipr"        "labeling"     "munsell"      "ade4"        
 [73] "segmented"    "gdtools"      "knitr"        "tidyr"        "crul"         "xml2"        
 [79] "bold"         "data.table"   "rredlist"     "rotl"         "ritis"        "worrms"      
 [85] "natserv"      "wikitaxa"     "phangorn"     "fansi"        "pillar"       "httr"        
 [91] "rvest"        "hoardr"       "curl"         "urltools"     "httpcode"     "mime"        
 [97] "codetools"    "systemfonts"  "vctrs"        "rappdirs"     "openssl"      "evaluate"    
[103] "highr"        "markdown"     "yaml"         "xfun"         "colorspace"   "quadprog"    
[109] "fastmatch"    "utf8"         "Rhdf5lib"     "solrium"      "rentrez"      "rncl"        
[115] "selectr"      "ellipsis"     "purrr"        "lifecycle"    "WikidataR"    "askpass"     
[121] "XML"          "progress"     "triebeard"    "backports"    "zeallot"      "WikipediR"   
[127] "sys"          "prettyunits"  "klaR"         "spdep"        "AlgDesign"    "combinat"    
[133] "questionr"    "sp"           "spData"       "sf"           "deldir"       "boot"        
[139] "LearnBayes"   "coda"         "expm"         "gmodels"      "gdata"        "shiny"       
[145] "miniUI"       "rstudioapi"   "classInt"     "htmltools"    "labelled"     "DBI"         
[151] "units"        "e1071"        "class"        "KernSmooth"   "gtools"       "haven"       
[157] "httpuv"       "xtable"       "sourcetools"  "later"        "promises"     "fastmap"     
[163] "forcats"      "Biobase"      "BiocGenerics" "Biostrings"   "multtest"     "S4Vectors"   
[169] "IRanges"      "XVector"      "survival"     "stats4"       "zlibbioc"    
</code></pre>
<div id="windows" class="section level3">
<h3>Windows</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="op">&gt;</span><span class="st"> </span><span class="kw">setwd</span>(<span class="st">&#39;path/to/workshop/folder&#39;</span>) <span class="co"># If the workshop folder is an rstudio project, this is not needed</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="op">&gt;</span><span class="st"> </span>pkg_info &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&#39;software/packages/package_info.csv&#39;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(pkg_info<span class="op">$</span>win_path, <span class="dt">repos =</span> <span class="ot">NULL</span>, <span class="dt">type =</span> <span class="st">&quot;win.binary&quot;</span>)</a></code></pre></div>
</div>
<div id="macos-and-linux" class="section level3">
<h3>macOS and Linux</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="op">&gt;</span><span class="st"> </span><span class="kw">setwd</span>(<span class="st">&#39;path/to/workshop/folder&#39;</span>) <span class="co"># If the workshop folder is an rstudio project, this is not needed</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="op">&gt;</span><span class="st"> </span>pkg_info &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&#39;software/packages/package_info.csv&#39;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(pkg_info<span class="op">$</span>source_path, <span class="dt">repos =</span> <span class="ot">NULL</span>, <span class="dt">type =</span> <span class="st">&quot;source&quot;</span>)</a></code></pre></div>
</div>
<div id="mac-osx-el-capitan" class="section level3">
<h3>Mac OSX (El Capitan)</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="op">&gt;</span><span class="st"> </span><span class="kw">setwd</span>(<span class="st">&#39;path/to/workshop/folder&#39;</span>) <span class="co"># If the workshop folder is an rstudio project, this is not needed</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="op">&gt;</span><span class="st"> </span>pkg_info &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&#39;software/packages/package_info.csv&#39;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="op">&gt;</span><span class="st"> </span><span class="kw">install.packages</span>(pkg_info<span class="op">$</span>osx_path, <span class="dt">repos =</span> <span class="ot">NULL</span>, <span class="dt">type =</span> <span class="st">&quot;mac.binary.el-capitan&quot;</span>)</a></code></pre></div>
</div>
</div>

<div class="footer_section">
  <div class="footer_content">
    <span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type">Metacoder's documentation</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://grunwaldlab.cgrb.oregonstate.edu/" property="cc:attributionName" rel="cc:attributionURL">The Grunwald lab and the USDA Horticultural Crops Research Unit</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/grunwaldlab/metacoder_documentation" rel="dct:source">https://github.com/grunwaldlab/metacoder_documentation</a>.
  </div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
