<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>workshop--07--diversity_stats.knit</title>

<script src="site_libs/header-attrs-2.17/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-84254038-1', 'auto');
  ga('send', 'pageview');

</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><div class="logo_frame"><span class="logo_helper"></span><img src="images/hex_sticker.png" height=50px /></div> metacoder</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="example.html">Example</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="workshop--00--introduction.html">Introduction</a>
    </li>
    <li>
      <a href="workshop--00--required_software.html">Required software</a>
    </li>
    <li>
      <a href="workshop--00--required_data.html">Required datasets</a>
    </li>
    <li>
      <a href="workshop--03--parsing.html">Getting data into R</a>
    </li>
    <li>
      <a href="workshop--04--manipulating.html">Manipulating taxonomic data</a>
    </li>
    <li>
      <a href="workshop--05--plotting.html">Plotting taxonomic data</a>
    </li>
    <li>
      <a href="workshop--06--quality_control.html">Data quality control</a>
    </li>
    <li>
      <a href="workshop--07--diversity_stats.html">Diversity and Ordination</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Publication
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404">PLOS Comp Bio paper</a>
    </li>
    <li>
      <a href="publication--05--16s_human_microbiome.html">Human microbiome</a>
    </li>
    <li>
      <a href="publication--06--tara_oceans.html">TARA Oceans</a>
    </li>
    <li>
      <a href="publication--01--silva.html">Digital PCR: SILVA</a>
    </li>
    <li>
      <a href="publication--02--rdp.html">Digital PCR: RDP</a>
    </li>
    <li>
      <a href="publication--03--greengenes.html">Digital PCR: Greengenes</a>
    </li>
    <li>
      <a href="publication--04--16s_database_comparision.html">Digital PCR: Comparing databases</a>
    </li>
    <li>
      <a href="publication--08--voting.html">Voting Geography</a>
    </li>
    <li>
      <a href="publication--09--gene_expression.html">Gene Expression</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Appendices
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="appendix--00--intro_to_r.html">Introduction to R</a>
    </li>
    <li>
      <a href="appendix--00--intro_to_rstudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="appendix--00--glossary.html">Glossary</a>
    </li>
    <li>
      <a href="appendix--00--presentation_page.html">Slide show</a>
    </li>
    <li>
      <a href="appendix--00--poster_page.html">Poster</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder">Metacoder source code</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder_documentation">Website source code</a>
    </li>
  </ul>
</li>
<li>
  <a href="faq.html">FAQ</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<div id="diversity-statistics" class="section level1">
<h1>Diversity statistics</h1>
<div id="load-example-data" class="section level2">
<h2>Load example data</h2>
<p>If you are starting the workshop at this section, or had problems
running code in a previous section, use the following to load the data
used in this section. You can download the “clean_data.Rdata” file
<a href="clean_data.Rdata" download="clean_data.Rdata">here</a>. If
<code>obj</code> and <code>sample_data</code> are already in your
environment, you can ignore this and proceed.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;clean_data.Rdata&quot;</span>)</span></code></pre></div>
</div>
<div id="measures-of-diversity" class="section level2">
<h2>Measures of diversity</h2>
<p>Diversity in the ecological sense is intuitively understood as the
complexity of a community of organisms. There are many ways to quantify
this complexity so that we can compare communities objectively. The two
main categories of methods are known as <strong>alpha diversity</strong>
and <strong>beta diversity</strong> <span class="citation">(Whittaker
1960)</span>. Alpha diversity measures the diversity within a single
sample and is generally based on the number and relative abundance of
taxa at some rank (e.g. species or OTUs). Beta diversity also uses the
number of relative abundance of taxa at some rank, but measures
variation between samples. In other words, an alpha diversity statistic
describes a single sample and a beta diversity statistic describes how
two samples compare.</p>
<p>The <code>vegan</code> package is the main tool set used for
calculating biological diversity statistics in R.</p>
</div>
<div id="alpha-within-sample-diversity" class="section level2">
<h2>Alpha (within sample) diversity</h2>
<p>Common alpha diversity statistics include:</p>
<ul>
<li><strong>Shannon</strong>: How difficult it is to predict the
identity of a randomly chosen individual.</li>
<li><strong>Simpson</strong>: The probability that two randomly chosen
individuals are the same species.</li>
<li><strong>Inverse Simpson</strong>: This is a bit confusing to think
about. Assuming a theoretical community where all species were equally
abundant, this would be the number of species needed to have the same
Simpson index value for the community being analyzed.</li>
</ul>
<p>There are also some diversity indexes that take into account the
taxonomic similarity of the species called “taxonomic diversity” and
“taxonomic distinctness”, but we will not go into those.</p>
<p>The <code>diversity</code> function from the <code>vegan</code>
package can be used to calculate the alpha diversity of a set of
samples. Like other vegan functions, it assumes that samples are in
rows, but they are in columns in our data, so we need to use the
<code>MARGIN = 2</code> option. We also need to exclude the taxon ID
column by subsetting the columns to only samples (i.e. all column
besides the first one). Since alpha diversity is a per-sample attribute,
we can just add this as a column to the sample data table:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>sample_data<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">diversity</span>(obj<span class="sc">$</span>data<span class="sc">$</span>otu_rarefied[, sample_data<span class="sc">$</span>SampleID],</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">MARGIN =</span> <span class="dv">2</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">index =</span> <span class="st">&quot;invsimpson&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(sample_data<span class="sc">$</span>alpha)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-2-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Adding this as a column to the sample data table makes it easy to
graph using <code>ggplot2</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sample_data, <span class="fu">aes</span>(<span class="at">x =</span> Site, <span class="at">y =</span> alpha)) <span class="sc">+</span> </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-3-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>We can use
<b><a href ="appendix--00--glossary.html#analysis_of_variance_(anova)_anchor">analysis
of variance (ANOVA)</a></b> to tell if at least one of the diversity
means is different from the rest.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>anova_result <span class="ot">&lt;-</span> <span class="fu">aov</span>(alpha <span class="sc">~</span> Site, sample_data)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(anova_result)</span></code></pre></div>
<pre class="r-output"><code>##              Df Sum Sq Mean Sq F value   Pr(&gt;F)    
## Site          2  80244   40122   14.69 8.42e-07 ***
## Residuals   289 789514    2732                     
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
</code></pre>
<p>That tells us that there is a difference, but does not tell us which
means are different. A
<b><a href ="appendix--00--glossary.html#tukey's_honest_significant_difference_(hsd)_anchor">Tukey’s
Honest Significant Difference (HSD)</a></b> test can do pairwise
comparisons of the means to find this out. We will use the
<code>HSD.test</code> function from the <code>agricolae</code> package
since it provides grouping codes that are useful for graphing.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(agricolae)</span></code></pre></div>
<pre><code>## Registered S3 methods overwritten by &#39;klaR&#39;:
##   method      from 
##   predict.rda vegan
##   print.rda   vegan
##   plot.rda    vegan</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>tukey_result <span class="ot">&lt;-</span> <span class="fu">HSD.test</span>(anova_result, <span class="st">&quot;Site&quot;</span>, <span class="at">group =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(tukey_result)</span></code></pre></div>
<pre class="r-output"><code>## $statistics
##    MSerror  Df     Mean       CV
##   2731.882 289 57.43052 91.00981
## 
## $parameters
##    test name.t ntr StudentizedRange alpha
##   Tukey   Site   3          3.33168  0.05
## 
## $means
##        alpha      std   r      Min       Max       Q25      Q50       Q75
## Jam 32.73475 31.23777  84 1.253188  95.52755  1.745386 33.65329  59.12646
## Mah 61.07740 48.69535 104 2.045294 154.05178 12.714725 56.75261 108.81348
## Sil 73.73023 67.13884 104 3.681719 240.22633 11.549273 34.04309 140.44495
## 
## $comparison
## NULL
## 
## $groups
##        alpha groups
## Sil 73.73023      a
## Mah 61.07740      a
## Jam 32.73475      b
## 
## attr(,"class")
## [1] "group"
</code></pre>
<p>Looking at the <code>tukey_result$groups</code> table it appears that
the alpha diversity of sites “Sil” and “Mah” might not be different, but
there is evidence that the diversity in site “Jam” is lower. We can add
this information to the graph using the
<code>tukey_result$groups$groups</code> codes:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>group_data <span class="ot">&lt;-</span> tukey_result<span class="sc">$</span>groups[<span class="fu">order</span>(<span class="fu">rownames</span>(tukey_result<span class="sc">$</span>groups)),]</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sample_data, <span class="fu">aes</span>(<span class="at">x =</span> Site, <span class="at">y =</span> alpha)) <span class="sc">+</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> <span class="fu">data.frame</span>(),</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">rownames</span>(group_data), <span class="at">y =</span> <span class="fu">max</span>(sample_data<span class="sc">$</span>alpha) <span class="sc">+</span> <span class="dv">1</span>, <span class="at">label =</span> group_data<span class="sc">$</span>groups),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">col =</span> <span class="st">&#39;black&#39;</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Alpha diversity&quot;</span>) <span class="sc">+</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Site&quot;</span>) <span class="sc">+</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">&quot;Alpha diversity index&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-6-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>So that takes care of comparing the alpha diversity of sites, but
there are other interesting groupings we can compare, such as the
genotype and the type of the sample (roots vs leaves). We could do the
above all over with minor modifications, but one of the benefits of
using a programming language is that you can create your own functions
to automate repeated tasks. We can generalize what we did above and put
it in a function like so:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>compare_alpha <span class="ot">&lt;-</span> <span class="cf">function</span>(sample_data, grouping_var) {</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calcuate alpha diversity</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  sample_data<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">diversity</span>(obj<span class="sc">$</span>data<span class="sc">$</span>otu_rarefied[, sample_data<span class="sc">$</span>SampleID],</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">MARGIN =</span> <span class="dv">2</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">index =</span> <span class="st">&quot;invsimpson&quot;</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Do ANOVA</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  sample_data<span class="sc">$</span>grouping <span class="ot">&lt;-</span> sample_data[[grouping_var]] <span class="co"># needed for how `aov` works</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  anova_result <span class="ot">&lt;-</span> <span class="fu">aov</span>(alpha <span class="sc">~</span> grouping, sample_data)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Do Tukey&#39;s HSD test</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  tukey_result <span class="ot">&lt;-</span> <span class="fu">HSD.test</span>(anova_result, <span class="st">&quot;grouping&quot;</span>, <span class="at">group =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plot result</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  group_data <span class="ot">&lt;-</span> tukey_result<span class="sc">$</span>groups[<span class="fu">order</span>(<span class="fu">rownames</span>(tukey_result<span class="sc">$</span>groups)),]</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  my_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sample_data, <span class="fu">aes</span>(<span class="at">x =</span> grouping, <span class="at">y =</span> alpha)) <span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="at">data =</span> <span class="fu">data.frame</span>(),</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>              <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">rownames</span>(group_data),</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>                  <span class="at">y =</span> <span class="fu">max</span>(sample_data<span class="sc">$</span>alpha) <span class="sc">+</span> <span class="dv">1</span>,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>                  <span class="at">label =</span> group_data<span class="sc">$</span>groups),</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>              <span class="at">col =</span> <span class="st">&#39;black&#39;</span>,</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>              <span class="at">size =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">&quot;Alpha diversity&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(grouping_var) <span class="sc">+</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">&quot;Alpha diversity index&quot;</span>)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return plot</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(my_plot)</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Using this function, we can compare plot the alpha diversities by
type of sample and genotype:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">compare_alpha</span>(sample_data, <span class="st">&quot;Type&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-8-1.png" width="480" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">compare_alpha</span>(sample_data, <span class="st">&quot;Genotype&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-8-2.png" width="480" style="display: block; margin: auto;" /></p>
<p>Looks like there is no difference in the alpha diversity between
genotypes, but a large difference between the diversity of roots and
leaves.</p>
<p>The <code>phyloseq</code> package (<span class="citation">McMurdie
and Holmes (2013)</span>) can be used to quickly plot a variety of alpha
diversity indexes per sample using the <code>plot_richness</code>
function. First we need to convert the <code>taxmap</code> object to a
<code>phyloseq</code> object, since all of the <code>phyloseq</code>
functions expect <code>phyloseq</code> objects.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(metacoder)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>ps_obj <span class="ot">&lt;-</span> <span class="fu">as_phyloseq</span>(obj,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">otu_table =</span> <span class="st">&quot;otu_rarefied&quot;</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">otu_id_col =</span> <span class="st">&quot;OTU_ID&quot;</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample_data =</span> sample_data,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample_id_col =</span> <span class="st">&quot;SampleID&quot;</span>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_richness</span>(ps_obj, <span class="at">color =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Site&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-9-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>Each dot is a sample and the error bars in some of the indexes are
the standard error.</p>
</div>
<div id="plotting-taxon-abundance-with-heat-trees"
class="section level2">
<h2>Plotting taxon abundance with heat trees</h2>
<p>Alpha diversity statistics capture the diversity of whole samples in
a single number, but to see the abundance of each taxon in a group of
samples (e.g., root samples), we need to use other techniques. Stacked
barcharts are typically used for this purpose, but we will be using heat
trees. First, we need to calculate the abundance of each taxon for a set
of samples from our OTU abundance information. We can use the
<code>calc_taxon_abund</code> function to do this, and then find the
mean abundance using a sample characteristic:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>tax_abund <span class="ot">&lt;-</span> <span class="fu">calc_taxon_abund</span>(obj, <span class="st">&quot;otu_props&quot;</span>)</span></code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Summing per-taxon counts from 292 columns for 489 taxa</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>type_abund <span class="ot">&lt;-</span> <span class="fu">calc_group_mean</span>(obj, <span class="st">&quot;tax_abund&quot;</span>,</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">cols =</span> sample_data<span class="sc">$</span>SampleID,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">groups =</span> sample_data<span class="sc">$</span>Type)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(obj<span class="sc">$</span>data<span class="sc">$</span>type_abund)</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #555555;'># A tibble: 489 × 3</span>
##    taxon_id     leaf    root
##    <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>   <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>
## <span style='color: #555555;'> 1</span> aad      1        1      
## <span style='color: #555555;'> 2</span> aaf      0.647    0.456  
## <span style='color: #555555;'> 3</span> aag      0.103    0.277  
## <span style='color: #555555;'> 4</span> aah      0.070<span style='text-decoration: underline;'>4</span>   0.015<span style='text-decoration: underline;'>5</span> 
## <span style='color: #555555;'> 5</span> aai      0.161    0.109  
## <span style='color: #555555;'> 6</span> aaj      0.004<span style='text-decoration: underline;'>38</span>  0.004<span style='text-decoration: underline;'>37</span>
## <span style='color: #555555;'> 7</span> aak      0.001<span style='text-decoration: underline;'>34</span>  0.026<span style='text-decoration: underline;'>6</span> 
## <span style='color: #555555;'> 8</span> aal      0.001<span style='text-decoration: underline;'>09</span>  0.031<span style='text-decoration: underline;'>8</span> 
## <span style='color: #555555;'> 9</span> aam      0.000<span style='text-decoration: underline;'>665</span> 0.013<span style='text-decoration: underline;'>5</span> 
## <span style='color: #555555;'>10</span> aan      0.005<span style='text-decoration: underline;'>71</span>  0.046<span style='text-decoration: underline;'>2</span> 
## <span style='color: #555555;'># … with 479 more rows</span>
</code></pre>
<p>Now we can use these per-taxon abundances to make heat trees of the
primary taxa present in leafs and roots:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>obj <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  metacoder<span class="sc">::</span><span class="fu">filter_taxa</span>(leaf <span class="sc">&gt;</span> <span class="fl">0.001</span>) <span class="sc">%&gt;%</span> <span class="co"># metacoder:: needed because of phyloseq::filter_taxa</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heat_tree</span>(<span class="at">node_label =</span> taxon_names,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_size =</span> leaf, </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color =</span> leaf, </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">layout =</span> <span class="st">&quot;da&quot;</span>, <span class="at">initial_layout =</span> <span class="st">&quot;re&quot;</span>, </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">title =</span> <span class="st">&quot;Taxa in leafs&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>obj <span class="sc">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  metacoder<span class="sc">::</span><span class="fu">filter_taxa</span>(root <span class="sc">&gt;</span> <span class="fl">0.001</span>) <span class="sc">%&gt;%</span> <span class="co"># metacoder:: needed because of phyloseq::filter_taxa</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heat_tree</span>(<span class="at">node_label =</span> taxon_names,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_size =</span> root, </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color =</span> root, </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">layout =</span> <span class="st">&quot;da&quot;</span>, <span class="at">initial_layout =</span> <span class="st">&quot;re&quot;</span>, </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">title =</span> <span class="st">&quot;Taxa in roots&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-12-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>Note that we needed to qualify <code>filter_taxa</code> with
<code>metacoder::</code>. This is because <code>phyloseq</code> is
loaded and it also has a function called <code>filter_taxa</code>. When
two functions from different packages have the same name, the function
from the package that was loaded last is called, unless
<code>package_name::</code> is added.</p>
</div>
<div id="beta-between-sample-diversity" class="section level2">
<h2>Beta (between sample) diversity</h2>
<p>Beta diversity is a way to quantify the difference between two
communities. There are many metrics that are used for this, but we will
only mention a few of the more popular ones. A few also incorporate
phylogenetic relatedness and require a phylogenetic tree of the
organisms in either community to be calculated.</p>
<div id="examples-of-indexes-used-with-presenceabsence-data"
class="section level4">
<h4>Examples of indexes used with presence/absence data:</h4>
<ul>
<li><strong>Sørensen</strong>: two times the number of species common to
both communities divided by the sum of the number of species in each
community.</li>
<li><strong>Jaccard</strong>: the number of species common to both
communities divided by the number of species in either community.</li>
<li><strong>Unifrac</strong>: The fraction of the phylogenetic tree
branch lengths shared by the two communities.</li>
</ul>
</div>
<div id="examples-of-indexes-used-with-count-data"
class="section level4">
<h4>Examples of indexes used with count data:</h4>
<ul>
<li><strong>Bray–Curtis</strong>: The sum of lesser counts for species
present in both communities divided by the sum of all counts in both
communities. This can be thought of as a quantitative version of the
Sørensen index.</li>
<li><strong>Weighted Unifrac</strong>: The fraction of the phylogenetic
tree branch lengths shared by the two communities, weighted by the
counts of organisms, so more abundant organisms have a greater
influence.</li>
</ul>
<p>The <code>vegan</code> function <code>vegdist</code> is used to
calculate the pairwise beta diversity indexes for a set of samples.
Since this is a pairwise comparison, the output is a triangular matrix.
In R, a <code>matrix</code> is like a <code>data.frame</code>, but all
of the same type (e.g. all numeric), and has some different
behavior.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>beta_dist <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(<span class="fu">t</span>(obj<span class="sc">$</span>data<span class="sc">$</span>otu_rarefied[, sample_data<span class="sc">$</span>SampleID]),</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">index =</span> <span class="st">&quot;bray&quot;</span>)</span></code></pre></div>
<p>Since <code>vegdist</code> does not have a <code>MARGIN</code> option
like <code>diversity</code>, we need to
<b><a href ="appendix--00--glossary.html#transpose_anchor">transpose</a></b>
the matrix with the <code>t</code> function.</p>
</div>
</div>
<div id="ordination" class="section level2">
<h2>Ordination</h2>
<p>The typical way beta diversity is plotted is using ordination.
Ordination is a way to display “high dimensional” data in a visible
number of dimensions (2 to 3). Our data is “high dimensional” because we
have many samples with many OTUs and the abundance of each OTU can be
considered a “dimension”. If we had only two species, we could make a
scatter plot of their abundance in each sample and get an idea of how
the samples differ. With thousands of species, this is not possible.
Instead, ordination is used to try to capture the information in many
dimensions by in a smaller number of new “artificial” dimensions.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mds <span class="ot">&lt;-</span> <span class="fu">metaMDS</span>(beta_dist)</span></code></pre></div>
<pre class="r-output"><code>## Run 0 stress 0.1131749 
## Run 1 stress 0.1144213 
## Run 2 stress 0.1165573 
## Run 3 stress 0.1149066 
## Run 4 stress 0.117722 
## Run 5 stress 0.1171036 
## Run 6 stress 0.1212251 
## Run 7 stress 0.115744 
## Run 8 stress 0.1162666 
## Run 9 stress 0.1156565 
## Run 10 stress 0.1135996 
## ... Procrustes: rmse 0.005077196  max resid 0.06732746 
## Run 11 stress 0.2062582 
## Run 12 stress 0.1143721 
## Run 13 stress 0.1135763 
## ... Procrustes: rmse 0.001682281  max resid 0.02776895 
## Run 14 stress 0.1571165 
## Run 15 stress 0.1149534 
## Run 16 stress 0.1135996 
## ... Procrustes: rmse 0.005077156  max resid 0.06732874 
## Run 17 stress 0.1138303 
## Run 18 stress 0.1163325 
## Run 19 stress 0.1138398 
## Run 20 stress 0.1136046 
## ... Procrustes: rmse 0.00307149  max resid 0.03814144 
## *** Best solution was not repeated -- monoMDS stopping criteria:
##      9: stress ratio &gt; sratmax
##     11: scale factor of the gradient &lt; sfgrmin
</code></pre>
<p>That transformed our beta diversity matrix into a set of coordinates
in two dimensions, which are intended to capture the differences in the
data. However, it is in a format specific to <code>vegan</code>, so we
will have to convert the data to a form that we can use for
plotting.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mds_data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mds<span class="sc">$</span>points)</span></code></pre></div>
<p>To use the sample data in the plotting, we can combine the coordinate
data with the sample data table:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>mds_data<span class="sc">$</span>SampleID <span class="ot">&lt;-</span> <span class="fu">rownames</span>(mds_data)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>mds_data <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(mds_data, sample_data)</span></code></pre></div>
<pre><code>## [1m[22mJoining, by = &quot;SampleID&quot;</code></pre>
<p>Now that we have the data in a format ggplot2 likes, we can plot it.
Lets plot our two new dimensions and color them by sample type
(i.e. leaves vs roots).</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_data, <span class="fu">aes</span>(<span class="at">x =</span> MDS1, <span class="at">y =</span> MDS2, <span class="at">color =</span> Type)) <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This shows that leaf and root samples are quite distinct, as we would
expect. We can also color them by Site:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_data, <span class="fu">aes</span>(<span class="at">x =</span> MDS1, <span class="at">y =</span> MDS2, <span class="at">color =</span> Site)) <span class="sc">+</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>It appears that within the leaf and root clusters, we “sub-clusters”
corresponding to site. Finally, lets look at genotype:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mds_data, <span class="fu">aes</span>(<span class="at">x =</span> MDS1, <span class="at">y =</span> MDS2, <span class="at">color =</span> Genotype)) <span class="sc">+</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>There is no discernible pattern there, suggesting plant genotype does
not correspond to community structure.</p>
<p>We can also do the above quickly in <code>phyloseq</code> using the
<code>ordinate</code> and <code>plot_ordination</code> functions. Lets
look at the differences between leaf and root samples again, but using a
difference index this time.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ps_ord <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(ps_obj, <span class="at">method =</span> <span class="st">&quot;NMDS&quot;</span>, <span class="at">distance =</span> <span class="st">&quot;jsd&quot;</span>)</span></code></pre></div>
<pre class="r-output"><code>## Run 0 stress 0.1069808 
## Run 1 stress 0.1106674 
## Run 2 stress 0.2097035 
## Run 3 stress 0.1093052 
## Run 4 stress 0.1069807 
## ... New best solution
## ... Procrustes: rmse 6.690557e-05  max resid 0.0006479441 
## ... Similar to previous best
## Run 5 stress 0.107159 
## ... Procrustes: rmse 0.004396367  max resid 0.07444733 
## Run 6 stress 0.1097307 
## Run 7 stress 0.1073264 
## ... Procrustes: rmse 0.004710866  max resid 0.07425749 
## Run 8 stress 0.1081149 
## Run 9 stress 0.1110496 
## Run 10 stress 0.1105936 
## Run 11 stress 0.1182278 
## Run 12 stress 0.1071952 
## ... Procrustes: rmse 0.001567977  max resid 0.02616063 
## Run 13 stress 0.1077171 
## Run 14 stress 0.1073263 
## ... Procrustes: rmse 0.004709749  max resid 0.07425838 
## Run 15 stress 0.1080435 
## Run 16 stress 0.1100731 
## Run 17 stress 0.1085896 
## Run 18 stress 0.1085896 
## Run 19 stress 0.107722 
## Run 20 stress 0.1073261 
## ... Procrustes: rmse 0.004707416  max resid 0.0742609 
## *** Best solution repeated 1 times
</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_ordination</span>(ps_obj, ps_ord, <span class="at">type =</span> <span class="st">&quot;samples&quot;</span>, <span class="at">color =</span> <span class="st">&quot;Type&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="differential-heat-trees" class="section level2">
<h2>Differential heat trees</h2>
<p>Beta diversity summarizes the difference between two samples in a
single number, but does not tell you why the samples are different. We
developed a plotting technique we call “differential heat trees” to
display differences in abundance of each taxon in two samples or groups
of samples. These are like the heat trees shown in the plotting section,
but colored with a diverging color scale the indicates which sample each
taxon is more abundant in and by how much. In this way, it is similar to
heat maps used for the same purpose in gene expression studies.</p>
<div id="comparing-taxon-abundance-in-two-groups"
class="section level3">
<h3>Comparing taxon abundance in two groups</h3>
<p>Before we can make a differential heat tree, we need to calculate the
difference in abundance for each taxon between groups of samples, such
as root vs leaf samples. There are many ways this can be done, ranging
from simple differences in mean read counts, to outputs from specialized
programs designed for microbiome data. We will be using a function in
<code>metacoder</code> called <code>compare_groups</code> to do the
comparisons. We can use the per-taxon read propotions we got with
<code>calc_taxon_abund</code> eariler for this. For each taxon, at every
rank, the <code>compare_groups</code> function compares two groups of
counts. We have to define which sample belongs to which groups using
<code>groups</code> option:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>diff_table <span class="ot">&lt;-</span> <span class="fu">compare_groups</span>(obj, <span class="at">data =</span> <span class="st">&quot;tax_abund&quot;</span>,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">cols =</span> sample_data<span class="sc">$</span>SampleID,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">groups =</span> sample_data<span class="sc">$</span>Type)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(obj<span class="sc">$</span>data<span class="sc">$</span>diff_table)</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #555555;'># A tibble: 489 × 7</span>
##    taxon_id treatment_1 treatment_2 log2_median_ratio median_diff  mean_diff wilcox_p_value
##    <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>    <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>                   <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>      <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>          <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>
## <span style='color: #555555;'> 1</span> aad      leaf        root                    0         0        0               3.21<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 2</span> aaf      leaf        root                    0.420     0.159    0.191           1.66<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-17</span>
## <span style='color: #555555;'> 3</span> aag      leaf        root                   -<span style='color: #BB0000;'>1.18</span>     -<span style='color: #BB0000;'>0.137</span>   -<span style='color: #BB0000;'>0.174</span>           4.06<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-43</span>
## <span style='color: #555555;'> 4</span> aah      leaf        root                    1.83      0.022<span style='text-decoration: underline;'>5</span>   0.054<span style='text-decoration: underline;'>9</span>          8.87<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-18</span>
## <span style='color: #555555;'> 5</span> aai      leaf        root                    0.448     0.038<span style='text-decoration: underline;'>5</span>   0.051<span style='text-decoration: underline;'>9</span>          1.94<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 2</span>
## <span style='color: #555555;'> 6</span> aaj      leaf        root                   -<span style='color: #BB0000;'>1.27</span>     -<span style='color: #BB0000;'>0.001</span><span style='color: #BB0000; text-decoration: underline;'>94</span>  0.000<span style='text-decoration: underline;'>011</span>5       3.32<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 9</span>
## <span style='color: #555555;'> 7</span> aak      leaf        root                   -<span style='color: #BB0000;'>6.72</span>     -<span style='color: #BB0000;'>0.024</span><span style='color: #BB0000; text-decoration: underline;'>9</span>  -<span style='color: #BB0000;'>0.025</span><span style='color: #BB0000; text-decoration: underline;'>2</span>          1.62<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-48</span>
## <span style='color: #555555;'> 8</span> aal      leaf        root                 -<span style='color: #BB0000;'>Inf</span>        -<span style='color: #BB0000;'>0.029</span><span style='color: #BB0000; text-decoration: underline;'>3</span>  -<span style='color: #BB0000;'>0.030</span><span style='color: #BB0000; text-decoration: underline;'>7</span>          8.43<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-51</span>
## <span style='color: #555555;'> 9</span> aam      leaf        root                 -<span style='color: #BB0000;'>Inf</span>        -<span style='color: #BB0000;'>0.012</span><span style='color: #BB0000; text-decoration: underline;'>9</span>  -<span style='color: #BB0000;'>0.012</span><span style='color: #BB0000; text-decoration: underline;'>8</span>          2.63<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-50</span>
## <span style='color: #555555;'>10</span> aan      leaf        root                   -<span style='color: #BB0000;'>4.66</span>     -<span style='color: #BB0000;'>0.042</span><span style='color: #BB0000; text-decoration: underline;'>3</span>  -<span style='color: #BB0000;'>0.040</span><span style='color: #BB0000; text-decoration: underline;'>5</span>          1.61<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-44</span>
## <span style='color: #555555;'># … with 479 more rows</span>
</code></pre>
<p>By default,
<b><a href ="appendix--00--glossary.html#wilcoxon_rank_sum_test_anchor">Wilcoxon
Rank Sum test</a></b> is used to test for statistical significance of
differences and various summary statistics are reported to measure the
size of the difference. Since we have done many independent tests (one
for each taxon), we need to
<b><a href ="appendix--00--glossary.html#multiple_comparison_corrections_anchor">correct
for multiple comparisions</a></b>. We will do that with a false
discovery rate (FDR) correction, but other types can be used as
well.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">mutate_obs</span>(obj, <span class="st">&quot;diff_table&quot;</span>,</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">wilcox_p_value =</span> <span class="fu">p.adjust</span>(wilcox_p_value, <span class="at">method =</span> <span class="st">&quot;fdr&quot;</span>))</span></code></pre></div>
<p>The most useful statistic for plotting is the log of ratio of median
abundances in the two groups, since it is centered on 0 and is symmetric
(e.g., a value of -4 is the same magnitude as 4). Lets set any
differences that are not significant to 0 so all differences shown on
the plot are significant.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>diff_table<span class="sc">$</span>log2_median_ratio[obj<span class="sc">$</span>data<span class="sc">$</span>diff_table<span class="sc">$</span>wilcox_p_value <span class="sc">&gt;</span> <span class="fl">0.05</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span></code></pre></div>
<p>Now we have all the info needed to make a differential heat tree. The
standard <code>heat_tree</code> function can be used, but a few things
need to be done to make an effective differential heat tree:</p>
<ul>
<li>A diverging color scale should be used, preferably with a neutral
color in the middle, like gray.</li>
<li>The interval of values displayed must be symmetric around zero so
the neutral middle color is centered on zero.</li>
<li>The node color should be set to the log of ratio of median
abundances in the two groups</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">heat_tree</span>(obj, </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_label =</span> taxon_names,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_size =</span> n_obs, <span class="co"># number of OTUs</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_color =</span> log2_median_ratio, <span class="co"># difference between groups</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_color_interval =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">10</span>), <span class="co"># symmetric interval</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_color_range =</span> <span class="fu">c</span>(<span class="st">&quot;cyan&quot;</span>, <span class="st">&quot;gray&quot;</span>, <span class="st">&quot;magenta&quot;</span>), <span class="co"># diverging colors</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_size_axis_label =</span> <span class="st">&quot;OTU count&quot;</span>,</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">node_color_axis_label =</span> <span class="st">&quot;Log 2 ratio of median counts&quot;</span>)</span></code></pre></div>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-24-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>That’s not too bad looking, but we can tweak it a bit to make it
better by changing the layout, adding a title, and filtering out some of
the taxa with odd names (depending on what the plot is for, you might
not want to remove these). Here <code>mutate_obs</code> is used to add a
temporary variable to the taxmap object that will contain the taxon
names with special characters like <code>[</code> removed.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>obj <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_obs</span>(<span class="st">&quot;cleaned_names&quot;</span>, <span class="fu">gsub</span>(taxon_names, <span class="at">pattern =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">[|</span><span class="sc">\\</span><span class="st">]&quot;</span>, <span class="at">replacement =</span> <span class="st">&quot;&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  metacoder<span class="sc">::</span><span class="fu">filter_taxa</span>(<span class="fu">grepl</span>(cleaned_names, <span class="at">pattern =</span> <span class="st">&quot;^[a-zA-Z]+$&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heat_tree</span>(<span class="at">node_label =</span> cleaned_names,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_size =</span> n_obs, <span class="co"># number of OTUs</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color =</span> log2_median_ratio, <span class="co"># difference between groups</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color_interval =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">10</span>, <span class="dv">10</span>), <span class="co"># symmetric interval</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color_range =</span> <span class="fu">c</span>(<span class="st">&quot;cyan&quot;</span>, <span class="st">&quot;gray&quot;</span>, <span class="st">&quot;magenta&quot;</span>), <span class="co"># diverging colors</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_size_axis_label =</span> <span class="st">&quot;OTU count&quot;</span>,</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color_axis_label =</span> <span class="st">&quot;Log 2 ratio of median counts&quot;</span>,</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">layout =</span> <span class="st">&quot;da&quot;</span>, <span class="at">initial_layout =</span> <span class="st">&quot;re&quot;</span>, <span class="co"># good layout for large trees</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">title =</span> <span class="st">&quot;leaf vs root samples&quot;</span>)</span></code></pre></div>
<pre><code>## Adding a new &quot;character&quot; vector of length 489.</code></pre>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-25-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>What color corresponds to each group depends on the order they were
given in the <code>compare_groups</code> function. Since “leaf” is
“treatment_1” in the “diff_table”, and “log2_median_ratio” is defined as
“log2(treatment_1 / treatment_2)”, when a taxon has more counts in leaf
samples, the ratio is positive, therefore taxa more abundant in leafs
are colored magenta in this case.</p>
</div>
<div id="comparing-taxon-abundance-with-more-than-2-groups"
class="section level3">
<h3>Comparing taxon abundance with more than 2 groups</h3>
<p>For pair-wise comparisons with more than two groups we have developed
a graphing technique we call a heat tree matrix, in which trees are made
for all pairs of sample groupings and arranged in a matrix with a larger
key tree that has the taxon names and a legend. The code to do this is
similar to the code for making a single differential heat tree like we
did above, but uses the <code>heat_tree_matrix</code> function. First we
need to use <code>compare_groups</code> to generate data for all
pair-wise comparisons for a grouping with more than two treatments. The
code below compares all the sites used in this study to eachother:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>diff_table <span class="ot">&lt;-</span> <span class="fu">compare_groups</span>(obj, <span class="at">data =</span> <span class="st">&quot;tax_abund&quot;</span>,</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">cols =</span> sample_data<span class="sc">$</span>SampleID,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">groups =</span> sample_data<span class="sc">$</span>Site)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(obj<span class="sc">$</span>data<span class="sc">$</span>diff_table)</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #555555;'># A tibble: 1,467 × 7</span>
##    taxon_id treatment_1 treatment_2 log2_median_ratio median_diff mean_diff wilcox_p_value
##    <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>    <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;chr&gt;</span>                   <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>       <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>     <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>          <span style='color: #555555; font-style: italic;'>&lt;dbl&gt;</span>
## <span style='color: #555555;'> 1</span> aad      Mah         Jam                    0        0          0              2.70<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 2</span> aaf      Mah         Jam                    0.187    0.062<span style='text-decoration: underline;'>1</span>    -<span style='color: #BB0000;'>0.093</span><span style='color: #BB0000; text-decoration: underline;'>7</span>         7.70<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 3</span> aag      Mah         Jam                   -<span style='color: #BB0000;'>0.202</span>   -<span style='color: #BB0000;'>0.027</span><span style='color: #BB0000; text-decoration: underline;'>7</span>    -<span style='color: #BB0000;'>0.033</span><span style='color: #BB0000; text-decoration: underline;'>0</span>         7.74<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 4</span> aah      Mah         Jam                    0.785    0.007<span style='text-decoration: underline;'>34</span>    0.007<span style='text-decoration: underline;'>96</span>        1.76<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 5</span> aai      Mah         Jam                    0.867    0.054<span style='text-decoration: underline;'>8</span>     0.107          7.29<span style='color: #555555;'>e</span><span style='color: #BB0000;'>-14</span>
## <span style='color: #555555;'> 6</span> aaj      Mah         Jam                   -<span style='color: #BB0000;'>0.050</span><span style='color: #BB0000; text-decoration: underline;'>9</span>  -<span style='color: #BB0000;'>0.000</span><span style='color: #BB0000; text-decoration: underline;'>090</span><span style='color: #BB0000;'>7</span>  0.000<span style='text-decoration: underline;'>561</span>       3.79<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 7</span> aak      Mah         Jam                   -<span style='color: #BB0000;'>0.391</span>   -<span style='color: #BB0000;'>0.002</span><span style='color: #BB0000; text-decoration: underline;'>95</span>   -<span style='color: #BB0000;'>0.001</span><span style='color: #BB0000; text-decoration: underline;'>83</span>        7.84<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 1</span>
## <span style='color: #555555;'> 8</span> aal      Mah         Jam                    0.180    0.001<span style='text-decoration: underline;'>62</span>    0.005<span style='text-decoration: underline;'>64</span>        7.09<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 2</span>
## <span style='color: #555555;'> 9</span> aam      Mah         Jam                   -<span style='color: #BB0000;'>0.672</span>   -<span style='color: #BB0000;'>0.003</span><span style='color: #BB0000; text-decoration: underline;'>00</span>   -<span style='color: #BB0000;'>0.002</span><span style='color: #BB0000; text-decoration: underline;'>16</span>        4.15<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 2</span>
## <span style='color: #555555;'>10</span> aan      Mah         Jam                    0.084<span style='text-decoration: underline;'>2</span>   0.000<span style='text-decoration: underline;'>995</span>   0.006<span style='text-decoration: underline;'>41</span>        3.20<span style='color: #555555;'>e</span><span style='color: #BB0000;'>- 3</span>
## <span style='color: #555555;'># … with 1,457 more rows</span>
</code></pre>
<p>We then need to correct for multiple comparisons and set
non-significant differences to zero like we did before:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">mutate_obs</span>(obj, <span class="st">&quot;diff_table&quot;</span>,</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">wilcox_p_value =</span> <span class="fu">p.adjust</span>(wilcox_p_value, <span class="at">method =</span> <span class="st">&quot;fdr&quot;</span>))</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>diff_table<span class="sc">$</span>log2_median_ratio[obj<span class="sc">$</span>data<span class="sc">$</span>diff_table<span class="sc">$</span>wilcox_p_value <span class="sc">&gt;</span> <span class="fl">0.05</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span></code></pre></div>
<p>Finally we call the <code>heat_tree_matrix</code> command with the
same options that would be used for a single tree.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>obj <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  metacoder<span class="sc">::</span><span class="fu">filter_taxa</span>(taxon_ranks <span class="sc">==</span> <span class="st">&quot;o&quot;</span>, <span class="at">supertaxa =</span> <span class="cn">TRUE</span>, <span class="at">reassign_obs =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_obs</span>(<span class="st">&quot;cleaned_names&quot;</span>, <span class="fu">gsub</span>(taxon_names, <span class="at">pattern =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">[|</span><span class="sc">\\</span><span class="st">]&quot;</span>, <span class="at">replacement =</span> <span class="st">&quot;&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  metacoder<span class="sc">::</span><span class="fu">filter_taxa</span>(<span class="fu">grepl</span>(cleaned_names, <span class="at">pattern =</span> <span class="st">&quot;^[a-zA-Z]+$&quot;</span>), <span class="at">reassign_obs =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heat_tree_matrix</span>(<span class="at">data =</span> <span class="st">&quot;diff_table&quot;</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_label =</span> cleaned_names,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_size =</span> n_obs, <span class="co"># number of OTUs</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_color =</span> log2_median_ratio, <span class="co"># difference between groups</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_color_trans =</span> <span class="st">&quot;linear&quot;</span>,</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_color_interval =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">3</span>), <span class="co"># symmetric interval</span></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">edge_color_interval =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3</span>, <span class="dv">3</span>), <span class="co"># symmetric interval</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_color_range =</span> <span class="fu">diverging_palette</span>(), <span class="co"># diverging colors</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_size_axis_label =</span> <span class="st">&quot;OTU count&quot;</span>,</span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>                   <span class="at">node_color_axis_label =</span> <span class="st">&quot;Log 2 ratio of median counts&quot;</span>,</span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>                   <span class="at">layout =</span> <span class="st">&quot;da&quot;</span>, <span class="at">initial_layout =</span> <span class="st">&quot;re&quot;</span>,</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>                   <span class="at">key_size =</span> <span class="fl">0.67</span>,</span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a>                   <span class="at">seed =</span> <span class="dv">2</span>)</span></code></pre></div>
<pre><code>## Adding a new &quot;character&quot; vector of length 247.</code></pre>
<p><img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-28-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>We have only compared three groups here, due to the nature of this
dataset, so this technique is not much better than 3 separate, full
sized graphs in this case, but with more groups it can be a uniquely
effective way to show lots of comparisons. See the end of the <a
href="02--quick_example.html">example analysis</a> for a better example
of this technique.</p>
</div>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<p>In these exercises, we will be using the <code>ps_obj</code> and
<code>obj</code> from the analysis above. If you did not run the code
above or had problems, run the following code to get the objects used.
You can download the “diversity_data.Rdata” file
<a href="diversity_data.Rdata" download="diversity_data.Rdata">here</a>.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;diversity_data.Rdata&quot;</span>)</span></code></pre></div>
<p><strong>1a)</strong> Look at the documentation for the
<code>plot_richness</code> function from <code>phyloseq</code>. Try to
make a plot using only the Simpson and inverse Simpson indexes, colored
by site and split up by sample type (leaf vs root).</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-31">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-31" class="collapse">
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_richness</span>(ps_obj, <span class="at">color =</span> <span class="st">&quot;Site&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Genotype&quot;</span>, <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">&quot;Simpson&quot;</span>, <span class="st">&quot;InvSimpson&quot;</span>))</span></code></pre></div>
<img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-31-1.png" width="480" style="display: block; margin: auto;" />
</div>
<p><br /></p>
<p><strong>1b)</strong> The Simpson and inverse Simpson indexes display
the same information in different ways (if you know one, you can
calculate the other). How do they differ?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-32">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-32" class="collapse">
The Simpson index is bound between 0 and 1, wheras the inverse Simpson
index is greater than zero. Also, the inverse Simpson index tends to
spread out higher diversity values more.
</div>
<p><br /></p>
<p><strong>2)</strong> Rarefaction and converting counts to proportions
are two ways of accounting for unequal sample depth. Although
proportions are more intuitive and easier to understand, why might
rarefaction be better when calculating diveristy indexes?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-33">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-33" class="collapse">
Some diversity indexes take into account the number of species. This
especially important for those that use presence/absence instead of
abundance, such as the Sørensen and Jaccard indexes. When read counts
are rarefied, some low counts go to zero, effectively reducing the
number of species in the sample, whereas converting to proportions will
never make a count go to zero, so will not remove the effect of unequal
sampling depth on these indexes.
</div>
<p><br /></p>
<p><strong>3a)</strong> Using the techniques presented in the section on
plotting the abundance of taxa, make a plot of taxon abundance for the
site encoded “Jam”.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-34">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-34" class="collapse">
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>tax_abund <span class="ot">&lt;-</span> <span class="fu">calc_taxon_abund</span>(obj, <span class="st">&quot;otu_props&quot;</span>)</span></code></pre></div>
<pre><code>## No `cols` specified, so using all numeric columns:
##    M1981P563, M1977P1709, M1980P502, M1981P606 ... M1955P747, M1978P342, M1955P778</code></pre>
<pre><code>## Summing per-taxon counts from 292 columns for 489 taxa</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">$</span>data<span class="sc">$</span>site_abund <span class="ot">&lt;-</span> <span class="fu">calc_group_mean</span>(obj, <span class="st">&quot;tax_abund&quot;</span>,</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">cols =</span> sample_data<span class="sc">$</span>SampleID,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">groups =</span> sample_data<span class="sc">$</span>Site)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>obj <span class="sc">%&gt;%</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  metacoder<span class="sc">::</span><span class="fu">filter_taxa</span>(Jam <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> <span class="co"># metacoder:: needed because of phyloseq::filter_taxa</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">heat_tree</span>(<span class="at">node_label =</span> taxon_names,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_size =</span> Jam, </span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_color =</span> Jam, </span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">layout =</span> <span class="st">&quot;da&quot;</span>, <span class="at">initial_layout =</span> <span class="st">&quot;re&quot;</span>, </span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">title =</span> <span class="st">&quot;Taxa in leafs&quot;</span>)</span></code></pre></div>
<img src="workshop--07--diversity_stats_files/figure-html/unnamed-chunk-34-1.png" width="480" style="display: block; margin: auto;" />
</div>
<p><br /></p>
<p><strong>3b)</strong> What are some of the most abundant families in
the site “Jam”?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-35">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-35" class="collapse">
Sphingobacteriaceae, Chitinophagaceae, Sphingomonadaceae,
Comamonadaceae, Microbacteriaceae, etc..
</div>
<p><br /></p>
<p><strong>3c)</strong> What is the most abundant phylum in the site
“Jam”?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-36">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-36" class="collapse">
Proteobacteria
</div>
<p><br /></p>
<p><strong>4a)</strong> The ordination analysis colored by genotype
showed no community differences corresponding to genotype. Try to use
<code>compare_groups</code> to see if any taxa are significantly
differentially abundant between genotypes to verify this result.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-37">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-37" class="collapse">
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>obj <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">compare_groups</span>(<span class="at">data =</span> <span class="st">&quot;tax_abund&quot;</span>,</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">cols =</span> sample_data<span class="sc">$</span>SampleID,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">groups =</span> sample_data<span class="sc">$</span>Genotype) <span class="sc">%&gt;%</span> </span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">wilcox_p_value =</span> <span class="fu">p.adjust</span>(wilcox_p_value, <span class="at">method =</span> <span class="st">&quot;fdr&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(wilcox_p_value <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #555555;'># A tibble: 0 × 7</span>
## <span style='color: #555555;'># … with 7 variables: taxon_id &lt;chr&gt;, treatment_1 &lt;chr&gt;, treatment_2 &lt;chr&gt;,</span>
## <span style='color: #555555;'>#   log2_median_ratio &lt;dbl&gt;, median_diff &lt;dbl&gt;, mean_diff &lt;dbl&gt;, wilcox_p_value &lt;dbl&gt;</span>
</code></pre>
</div>
<p><br /></p>
<p><strong>4b)</strong> How many taxa have significant differences
between genotypes?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-38">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-38" class="collapse">
None
</div>
<p><br /></p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="unnumbered">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-mcmurdie2013phyloseq" class="csl-entry">
McMurdie, Paul J, and Susan Holmes. 2013. <span>“Phyloseq: An r Package
for Reproducible Interactive Analysis and Graphics of Microbiome Census
Data.”</span> <em>PloS One</em> 8 (4): e61217. <a
href="https://doi.org/10.1371/journal.pone.0061217">https://doi.org/10.1371/journal.pone.0061217</a>.
</div>
<div id="ref-whittaker1960vegetation" class="csl-entry">
Whittaker, Robert Harding. 1960. <span>“Vegetation of the Siskiyou
Mountains, Oregon and California.”</span> <em>Ecological Monographs</em>
30 (3): 279–338.
</div>
</div>
</div>
</div>

<div class="footer_section">
  <div class="footer_content">
    <span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type">Metacoder's documentation</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://grunwaldlab.cgrb.oregonstate.edu/" property="cc:attributionName" rel="cc:attributionURL">The Grunwald lab and the USDA Horticultural Crops Research Unit</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/grunwaldlab/metacoder_documentation" rel="dct:source">https://github.com/grunwaldlab/metacoder_documentation</a>.
  </div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
