<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title></title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-84254038-1', 'auto');
  ga('send', 'pageview');

</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 61px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 66px;
  margin-top: -66px;
}

.section h2 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h3 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h4 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h5 {
  padding-top: 66px;
  margin-top: -66px;
}
.section h6 {
  padding-top: 66px;
  margin-top: -66px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><div class="logo_frame"><span class="logo_helper"></span><img src="images/hex_sticker.png" height=50px /></div> metacoder</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="example.html">Example</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="workshop--00--introduction.html">Introduction</a>
    </li>
    <li>
      <a href="workshop--00--required_software.html">Required software</a>
    </li>
    <li>
      <a href="workshop--00--required_data.html">Required datasets</a>
    </li>
    <li>
      <a href="workshop--03--parsing.html">Getting data into R</a>
    </li>
    <li>
      <a href="workshop--04--manipulating.html">Manipulating taxonomic data</a>
    </li>
    <li>
      <a href="workshop--05--plotting.html">Plotting taxonomic data</a>
    </li>
    <li>
      <a href="workshop--06--quality_control.html">Data quality control</a>
    </li>
    <li>
      <a href="workshop--07--diversity_stats.html">Diversity and Ordination</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Publication
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404">PLOS Comp Bio paper</a>
    </li>
    <li>
      <a href="publication--05--16s_human_microbiome.html">Human microbiome</a>
    </li>
    <li>
      <a href="publication--06--tara_oceans.html">TARA Oceans</a>
    </li>
    <li>
      <a href="publication--01--silva.html">Digital PCR: SILVA</a>
    </li>
    <li>
      <a href="publication--02--rdp.html">Digital PCR: RDP</a>
    </li>
    <li>
      <a href="publication--03--greengenes.html">Digital PCR: Greengenes</a>
    </li>
    <li>
      <a href="publication--04--16s_database_comparision.html">Digital PCR: Comparing databases</a>
    </li>
    <li>
      <a href="publication--08--voting.html">Voting Geography</a>
    </li>
    <li>
      <a href="publication--09--gene_expression.html">Gene Expression</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Appendices
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="appendix--00--intro_to_r.html">Introduction to R</a>
    </li>
    <li>
      <a href="appendix--00--intro_to_rstudio.html">Introduction to RStudio</a>
    </li>
    <li>
      <a href="appendix--00--glossary.html">Glossary</a>
    </li>
    <li>
      <a href="appendix--00--presentation_page.html">Slide show</a>
    </li>
    <li>
      <a href="appendix--00--poster_page.html">Poster</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder">Metacoder source code</a>
    </li>
    <li>
      <a href="https://github.com/grunwaldlab/metacoder_documentation">Website source code</a>
    </li>
  </ul>
</li>
<li>
  <a href="faq.html">FAQ</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">




</div>


<div id="getting-data-into-r" class="section level1">
<h1>Getting data into R</h1>
<div id="typical-format-for-microbiome-data" class="section level2">
<h2>Typical format for microbiome data</h2>
<p>Most <b><a href ="appendix--00--glossary.html#pipelines_anchor">pipelines</a></b> for processing high-throughput <b><a href ="appendix--00--glossary.html#amplicon_anchor">amplicon</a></b> data, such as mothur, QIIME, and dada2, result in a matrix of read counts. One dimension of this matrix (i.e.Â the rows or the columns) consists of <b><a href ="appendix--00--glossary.html#operational_taxonomic_units_(otus)_anchor">Operational Taxonomic Units (OTUs)</a></b>, <b><a href ="appendix--00--glossary.html#phylotypes_anchor">phylotypes</a></b>, or <b><a href ="appendix--00--glossary.html#exact_sequence_variants_(esvs)_anchor">exact sequence variants (ESVs)</a></b> (all ways to âbinâ similar read sequences). The other dimension consists of samples. Different tools will expect/output different orientations of the matrix, but, in our case, columns are samples and rows are OTUs. Sometimes the <b><a href ="appendix--00--glossary.html#operational_taxonomic_units_(otus)_anchor">OTU</a></b> data and the abundance matrix are two separate tables. There is usually another table with sample information in rows. This makes it easy to add lots of additional sample data columns that can be used to subset the data. Each sample and OTU will have an unique ID.</p>
<p><img src="figure_sources/typical_input_format.png" width="780" style="display: block; margin: auto;" /></p>
</div>
<div id="importing-data-into-r" class="section level2">
<h2>Importing data into R</h2>
<p>Importing data into R can be quite easy if the data is formatted well, but can be a very frustrating experience otherwise. An example of well-formatted data is <b><a href ="appendix--00--glossary.html#comma-delimited_text_file_anchor">.csv (comma-separated value)</a></b> or <b><a href ="appendix--00--glossary.html#tab-delimited_text_file_anchor">.tsv (tab-separated value)</a></b> files, each with a single table and no additional comments or formatting (e.g.Â merged cells). Either of these formats might also have a .txt extension (the extension does not really matter; its for humans, not computers). For more information on correct data formatting, see the <a href="http://grunwaldlab.github.io/Reproducible-science-in-R/03--Data_formatting.html">data formating section</a> of our <a href="http://grunwaldlab.github.io/Reproducible-science-in-R/index.html">guide for reporducible research</a>. You should always import the raw output data whenever possible and avoid any âmanualâ (i.e.Â non-scripted) modification of the data, especially in programs like Excel, which are known to mangle data from time to time (<span class="citation">Zeeberg et al. (2004)</span>).</p>
<p>Throughout this workshop, we will be using data from <span class="citation">Wagner et al. (2016)</span>, a study on the effects of plant age, genotype, and environment on the bacterial microbiome of <a href="https://en.wikipedia.org/wiki/Boechera_stricta"><em>Boechera stricta</em></a>, a perennial herb in the mustard family. Here is a photo of <em>Boechera stricta</em> taken by Mary Ellen Harte:</p>
<p><img src="images/boechera_stricta.jpg" width="375" style="display: block; margin: auto;" /></p>
<p><span class="citation">Wagner et al. (2016)</span> released their raw data with the article and it is available <a href="http://datadryad.org/resource/doi:10.5061/dryad.g60r3">here</a> on <a href="http://datadryad.org/">dryad</a>. This is a great example of how to share your raw data!</p>
<p>There are many <b><a href ="appendix--00--glossary.html#function_anchor">functions</a></b> commonly used to read tabular data, including the <b><a href ="appendix--00--glossary.html#base_r_anchor">base R</a></b> ones like <code>read.table</code> and <code>read.csv</code>, but we will be using functions from the new <a href="http://readr.tidyverse.org/articles/readr.html"><code>readr</code> package</a>, which returns <b><a href ="appendix--00--glossary.html#tibble_anchor">tibbles</a></b> instead of <code>data.frame</code>s (A âtableâ in R).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(readr) <span class="co"># Loads the readr package so we can use `read_tsv`</span></code></pre></div>
<p>Tibbles are a type of <code>data.frame</code> with some fancier printing and more consistent behavior. Click <a href="otuTable97.txt.bz2" download="otuTable97.txt.bz2">here</a> to download the OTU table. Lets read in the raw OTU table first:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">otu_data &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;otuTable97.txt.bz2&quot;</span>) <span class="co"># You might need to change the path to the file</span>
<span class="kw">print</span>(otu_data) <span class="co"># You can also enter just `otu_data` to print it</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 47,806 x 1,699</span><span>
##    OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38 M1551P90 M1551P71
##     </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span>      1          0        0        0        0        0        0        0        0        0
## </span><span style='color: #BCBCBC;'> 2</span><span>      2         41       22        4      726   </span><span style='text-decoration: underline;'>112</span><span>492        2      413        2        1
## </span><span style='color: #BCBCBC;'> 3</span><span>      3         67       65       12    </span><span style='text-decoration: underline;'>13</span><span>514        1        4    </span><span style='text-decoration: underline;'>13</span><span>314       70     </span><span style='text-decoration: underline;'>5</span><span>929
## </span><span style='color: #BCBCBC;'> 4</span><span>      4       </span><span style='text-decoration: underline;'>3</span><span>229    </span><span style='text-decoration: underline;'>13</span><span>679     </span><span style='text-decoration: underline;'>1</span><span>832      951      113     </span><span style='text-decoration: underline;'>2</span><span>496      567     </span><span style='text-decoration: underline;'>2</span><span>428     </span><span style='text-decoration: underline;'>2</span><span>156
## </span><span style='color: #BCBCBC;'> 5</span><span>      5       </span><span style='text-decoration: underline;'>1</span><span>200       92     </span><span style='text-decoration: underline;'>3</span><span>530     </span><span style='text-decoration: underline;'>2</span><span>008        0      183     </span><span style='text-decoration: underline;'>2</span><span>087      292     </span><span style='text-decoration: underline;'>1</span><span>058
## </span><span style='color: #BCBCBC;'> 6</span><span>      6        219     </span><span style='text-decoration: underline;'>1</span><span>980     </span><span style='text-decoration: underline;'>1</span><span>200      499        1      781      214     </span><span style='text-decoration: underline;'>2</span><span>273      171
## </span><span style='color: #BCBCBC;'> 7</span><span>      7        485     </span><span style='text-decoration: underline;'>5</span><span>123      755     </span><span style='text-decoration: underline;'>4</span><span>080      443     </span><span style='text-decoration: underline;'>1</span><span>278     </span><span style='text-decoration: underline;'>2</span><span>193      401     </span><span style='text-decoration: underline;'>5</span><span>320
## </span><span style='color: #BCBCBC;'> 8</span><span>      8        840     </span><span style='text-decoration: underline;'>7</span><span>079     </span><span style='text-decoration: underline;'>3</span><span>760       22        0     </span><span style='text-decoration: underline;'>2</span><span>699       22     </span><span style='text-decoration: underline;'>2</span><span>870       32
## </span><span style='color: #BCBCBC;'> 9</span><span>      9         40       82       91      881        1      449     </span><span style='text-decoration: underline;'>1</span><span>121       90     </span><span style='text-decoration: underline;'>2</span><span>283
## </span><span style='color: #BCBCBC;'>10</span><span>     10         11       79      277     </span><span style='text-decoration: underline;'>2</span><span>879        1        0     </span><span style='text-decoration: underline;'>6</span><span>811       14      243
## </span><span style='color: #949494;'># ... with 47,796 more rows, and 1,689 more variables: M1551P12 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P84 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1551P48 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P4 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P52 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P3 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P15 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P31 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1551P75 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P88 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, â¦</span><span>
</span></code></pre>
<p>This is a big data set, with 47,806 rows (OTUs) and 1,699 columns (1,698 samples and an OTU ID). If your computer cannot load this file, donât worry, we will provide a subset later for the rest of the workshop.</p>
<p>In this data set, the taxonomic classifications for the OTUs are in a different file. This information could have been included as additional columns in the OTU table and often is in other data sets. Click <a href="taxAssignments97.txt" download="taxAssignments97.txt">here</a> to download the taxonomic classifications table.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tax_data &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;taxAssignments97.txt&quot;</span>)
<span class="kw">print</span>(tax_data) <span class="co"># You can also enter `tax_data` to print it</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 47,806 x 8</span><span>
##    `OTU ID`  taxonomy                       Kingdom  Phylum  Class    Order    Family    Confidence
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                          </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>          </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> OTU_1     Unassigned                     Unassigâ¦ Unassiâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'> 2</span><span> OTU_10    Root;k__Bacteria;p__Bacteroidâ¦ Bacteria Bacterâ¦ Sphingoâ¦ Sphingoâ¦ Sphingobâ¦       1   
## </span><span style='color: #BCBCBC;'> 3</span><span> OTU_100   Root;k__Bacteria;p__Cyanobactâ¦ Bacteria Cyanobâ¦ Chloropâ¦ Chloropâ¦ </span><span style='color: #BB0000;'>NA</span><span>              1   
## </span><span style='color: #BCBCBC;'> 4</span><span> OTU_1000  Root;k__Bacteria;p__Actinobacâ¦ Bacteria Actinoâ¦ Actinobâ¦ Actinomâ¦ Actinosyâ¦       0.67
## </span><span style='color: #BCBCBC;'> 5</span><span> OTU_10000 Unassigned                     Unassigâ¦ Unassiâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'> 6</span><span> OTU_10001 Root;k__Bacteria;p__Chlamydiaâ¦ Bacteria Chlamyâ¦ Chlamydâ¦ Chlamydâ¦ Parachlaâ¦       1   
## </span><span style='color: #BCBCBC;'> 7</span><span> OTU_10002 Root;k__Bacteria;p__Proteobacâ¦ Bacteria Proteoâ¦ Alphaprâ¦ </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>              1   
## </span><span style='color: #BCBCBC;'> 8</span><span> OTU_10003 Unassigned                     Unassigâ¦ Unassiâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'> 9</span><span> OTU_10004 Unassigned                     Unassigâ¦ Unassiâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'>10</span><span> OTU_10005 Root;k__Bacteria;p__Cyanobactâ¦ Bacteria Cyanobâ¦ 4C0d-2   MLE1-12  </span><span style='color: #BB0000;'>NA</span><span>              1   
## </span><span style='color: #949494;'># ... with 47,796 more rows</span><span>
</span></code></pre>
<p>Although these data are very well-formatted compared to most, there are still a few issues. The âOTU IDâ column contains a space in the name (hence the back ticks), which makes it a bit more annoying to work with in R. More importantly, the OTU IDs in the taxonomy table are prefixed with âOTU_â and those in the OTU table are not, so we have to remove that prefix to make the two match up. The functions <code>sub</code> and <code>gsub</code> are used to search and replace parts of text; <code>sub</code> replaces only the first match and <code>gsub</code> replaces all matches. Replacing with nothing (<code>&quot;&quot;</code>) effectively searches and deletes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tax_data<span class="op">$</span><span class="st">`</span><span class="dt">OTU ID</span><span class="st">`</span> &lt;-<span class="st"> </span><span class="kw">sub</span>(tax_data<span class="op">$</span><span class="st">`</span><span class="dt">OTU ID</span><span class="st">`</span>, <span class="co"># ` are needed because of the space</span>
                         <span class="dt">pattern =</span> <span class="st">&quot;OTU_&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)
<span class="kw">print</span>(tax_data) </code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 47,806 x 8</span><span>
##    `OTU ID` taxonomy                       Kingdom  Phylum   Class    Order    Family    Confidence
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                          </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>          </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> 1        Unassigned                     Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'> 2</span><span> 10       Root;k__Bacteria;p__Bacteroidâ¦ Bacteria Bacteroâ¦ Sphingoâ¦ Sphingoâ¦ Sphingobâ¦       1   
## </span><span style='color: #BCBCBC;'> 3</span><span> 100      Root;k__Bacteria;p__Cyanobactâ¦ Bacteria Cyanobaâ¦ Chloropâ¦ Chloropâ¦ </span><span style='color: #BB0000;'>NA</span><span>              1   
## </span><span style='color: #BCBCBC;'> 4</span><span> 1000     Root;k__Bacteria;p__Actinobacâ¦ Bacteria Actinobâ¦ Actinobâ¦ Actinomâ¦ Actinosyâ¦       0.67
## </span><span style='color: #BCBCBC;'> 5</span><span> 10000    Unassigned                     Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'> 6</span><span> 10001    Root;k__Bacteria;p__Chlamydiaâ¦ Bacteria Chlamydâ¦ Chlamydâ¦ Chlamydâ¦ Parachlaâ¦       1   
## </span><span style='color: #BCBCBC;'> 7</span><span> 10002    Root;k__Bacteria;p__Proteobacâ¦ Bacteria Proteobâ¦ Alphaprâ¦ </span><span style='color: #BB0000;'>NA</span><span>       </span><span style='color: #BB0000;'>NA</span><span>              1   
## </span><span style='color: #BCBCBC;'> 8</span><span> 10003    Unassigned                     Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'> 9</span><span> 10004    Unassigned                     Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassigâ¦ Unassignâ¦       1   
## </span><span style='color: #BCBCBC;'>10</span><span> 10005    Root;k__Bacteria;p__Cyanobactâ¦ Bacteria Cyanobaâ¦ 4C0d-2   MLE1-12  </span><span style='color: #BB0000;'>NA</span><span>              1   
## </span><span style='color: #949494;'># ... with 47,796 more rows</span><span>
</span></code></pre>
<p>Although we could proceed with the analysis using separate OTU and taxonomy tables, lets combine them to simplify things. Since the rows are in different order, we need to combine (aka âjoinâ) them based on their OTU ID. We will use the <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/dplyr.html"><code>dplyr</code> package</a> for this.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr) <span class="co"># Loads the dplyr package so we can use `left_join`</span>
tax_data<span class="op">$</span><span class="st">`</span><span class="dt">OTU ID</span><span class="st">`</span> &lt;-<span class="st"> </span><span class="kw">as.character</span>(tax_data<span class="op">$</span><span class="st">`</span><span class="dt">OTU ID</span><span class="st">`</span>) <span class="co"># Must be same type for join to work</span>
otu_data<span class="op">$</span>OTU_ID &lt;-<span class="st"> </span><span class="kw">as.character</span>(otu_data<span class="op">$</span>OTU_ID) <span class="co"># Must be same type for join to work</span>
otu_data &lt;-<span class="st"> </span><span class="kw">left_join</span>(otu_data, tax_data,
                      <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;OTU_ID&quot;</span> =<span class="st"> &quot;OTU ID&quot;</span>)) <span class="co"># identifies cols with shared IDs</span>
<span class="kw">print</span>(otu_data)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 47,806 x 1,706</span><span>
##    OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38 M1551P90 M1551P71
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> 1               0        0        0        0        0        0        0        0        0
## </span><span style='color: #BCBCBC;'> 2</span><span> 2              41       22        4      726   </span><span style='text-decoration: underline;'>112</span><span>492        2      413        2        1
## </span><span style='color: #BCBCBC;'> 3</span><span> 3              67       65       12    </span><span style='text-decoration: underline;'>13</span><span>514        1        4    </span><span style='text-decoration: underline;'>13</span><span>314       70     </span><span style='text-decoration: underline;'>5</span><span>929
## </span><span style='color: #BCBCBC;'> 4</span><span> 4            </span><span style='text-decoration: underline;'>3</span><span>229    </span><span style='text-decoration: underline;'>13</span><span>679     </span><span style='text-decoration: underline;'>1</span><span>832      951      113     </span><span style='text-decoration: underline;'>2</span><span>496      567     </span><span style='text-decoration: underline;'>2</span><span>428     </span><span style='text-decoration: underline;'>2</span><span>156
## </span><span style='color: #BCBCBC;'> 5</span><span> 5            </span><span style='text-decoration: underline;'>1</span><span>200       92     </span><span style='text-decoration: underline;'>3</span><span>530     </span><span style='text-decoration: underline;'>2</span><span>008        0      183     </span><span style='text-decoration: underline;'>2</span><span>087      292     </span><span style='text-decoration: underline;'>1</span><span>058
## </span><span style='color: #BCBCBC;'> 6</span><span> 6             219     </span><span style='text-decoration: underline;'>1</span><span>980     </span><span style='text-decoration: underline;'>1</span><span>200      499        1      781      214     </span><span style='text-decoration: underline;'>2</span><span>273      171
## </span><span style='color: #BCBCBC;'> 7</span><span> 7             485     </span><span style='text-decoration: underline;'>5</span><span>123      755     </span><span style='text-decoration: underline;'>4</span><span>080      443     </span><span style='text-decoration: underline;'>1</span><span>278     </span><span style='text-decoration: underline;'>2</span><span>193      401     </span><span style='text-decoration: underline;'>5</span><span>320
## </span><span style='color: #BCBCBC;'> 8</span><span> 8             840     </span><span style='text-decoration: underline;'>7</span><span>079     </span><span style='text-decoration: underline;'>3</span><span>760       22        0     </span><span style='text-decoration: underline;'>2</span><span>699       22     </span><span style='text-decoration: underline;'>2</span><span>870       32
## </span><span style='color: #BCBCBC;'> 9</span><span> 9              40       82       91      881        1      449     </span><span style='text-decoration: underline;'>1</span><span>121       90     </span><span style='text-decoration: underline;'>2</span><span>283
## </span><span style='color: #BCBCBC;'>10</span><span> 10             11       79      277     </span><span style='text-decoration: underline;'>2</span><span>879        1        0     </span><span style='text-decoration: underline;'>6</span><span>811       14      243
## </span><span style='color: #949494;'># ... with 47,796 more rows, and 1,696 more variables: M1551P12 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P84 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1551P48 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P4 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P52 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P3 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P15 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P31 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1551P75 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P88 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, â¦</span><span>
</span></code></pre>
<p>There are so many columns that all of them are not shown in the print out, but we can verify that they are there by looking at the last 10 column names:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tail</span>(<span class="kw">colnames</span>(otu_data), <span class="dt">n =</span> <span class="dv">10</span>) <span class="co"># `tail` returns the last n elements</span></code></pre></div>
<pre class="r-output"><code>##  [1] "M1958P1043" "M1691P1526" "M1691P1557" "taxonomy"   "Kingdom"    "Phylum"     "Class"     
##  [8] "Order"      "Family"     "Confidence"
</code></pre>
<p>Next, lets load the sample data. Click <a href="SMD.txt" download="SMD.txt">here</a> to download the sample data table.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_data &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;SMD.txt&quot;</span>,
                        <span class="dt">col_types =</span> <span class="st">&quot;cccccccccccccccc&quot;</span>) <span class="co"># each &quot;c&quot; means a column of &quot;character&quot;</span>
<span class="kw">print</span>(sample_data) <span class="co"># You can also enter `sample_data` to print it</span></code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 1,698 x 16</span><span>
##    SampleID Name  Plant_ID Type  Experiment Cohort Harvested Age   Site  Treatment Line  Genotype
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>      </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>  </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>     </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>   
## </span><span style='color: #BCBCBC;'> 1</span><span> M1024P1â¦ R_02â¦ R_026    root  fieldBCMA  2008   2011      3     LTM   field     26    ril     
## </span><span style='color: #BCBCBC;'> 2</span><span> M1024P1â¦ R_07â¦ R_073    root  fieldBCMA  2008   2011      3     LTM   field     73    ril     
## </span><span style='color: #BCBCBC;'> 3</span><span> M1024P1â¦ R_08â¦ R_088    root  fieldBCMA  2009   2011      2     LTM   field     88    ril     
## </span><span style='color: #BCBCBC;'> 4</span><span> M1024P1â¦ R_15â¦ R_156    root  fieldBCMA  2009   2011      2     LTM   field     156   ril     
## </span><span style='color: #BCBCBC;'> 5</span><span> M1955P8â¦ 1_A_â¦ 1_A_1    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   par1â¦ PAR     
## </span><span style='color: #BCBCBC;'> 6</span><span> M1956P8â¦ 1_A_â¦ 1_A_12   root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   mah3â¦ MAH     
## </span><span style='color: #BCBCBC;'> 7</span><span> M1957P9â¦ 1_A_â¦ 1_A_3    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   silLâ¦ SIL     
## </span><span style='color: #BCBCBC;'> 8</span><span> M1956P8â¦ 1_A_â¦ 1_A_4    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   par1â¦ PAR     
## </span><span style='color: #BCBCBC;'> 9</span><span> M1957P9â¦ 1_A_â¦ 1_A_7    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   par9â¦ PAR     
## </span><span style='color: #BCBCBC;'>10</span><span> M1957P9â¦ 1_A_â¦ 1_A_8    root  ecoGH      </span><span style='color: #BB0000;'>NA</span><span>     2011      </span><span style='color: #BB0000;'>NA</span><span>    Duke  MAHsoil   mil5â¦ MIL     
## </span><span style='color: #949494;'># ... with 1,688 more rows, and 4 more variables: Block </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, oldPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>, newPlate </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span style='color: #949494;'>,
## #   Analysis </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
</span></code></pre>
<p>Note how the number of sample columns in <code>otu_data</code> is equal to the number of rows in <code>sample_data</code> and the columns names of <code>otu_data</code> appear in the âSampleIDâ column. This means that the contents of <code>sample_data$SampleID</code> can be used to subset columns in the OTU table.</p>
</div>
<div id="converting-to-the-taxmap-format" class="section level2">
<h2>Converting to the <code>taxmap</code> format</h2>
<p>Although our data is now in R, it is not in a format that is specialized for community abundance data; all R knows is that you have a few big tables. Different R packages for community (e.g.Â microbiome) analysis expect data in different formats or <b><a href ="appendix--00--glossary.html#class_anchor">classes</a></b>. A class, in programming jargon, is a defined way to store data plus some functions designed to interact with that data. When you format a specific data set in this way, we call it an <b><a href ="appendix--00--glossary.html#object_anchor">object</a></b> or an âinstanceâ of the class. Many R packages implement their own classes and functions to convert data to their format, whereas some packages use the classes defined in other packages. There are a few options for how to store an abundance matrix classified by a taxonomy in R (e.g. <code>phyloseq</code> objects), but we will be using classes defined in the <code>taxa</code> package here. The goal of the <code>taxa</code> package is to provide an all-purpose standard way of manipulation any type of information assigned to a taxonomy. <code>Taxa</code> provides a set of flexible <b><a href ="appendix--00--glossary.html#parsing_anchor">parsers</a></b> that should be able to read nearly any format, given the correct settings. You can read more about parsing taxonomic data with <code>taxa</code> here: <a href="https://github.com/ropensci/taxa#parsing-data" class="uri">https://github.com/ropensci/taxa#parsing-data</a>. The taxonomic data we appended to the abundance matrix has the following form:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(otu_data<span class="op">$</span>taxonomy, <span class="dv">10</span>)</code></pre></div>
<pre class="r-output"><code>##  [1] "Unassigned"                                                                                        
##  [2] "Root;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rickettsiales;f__mitochondria"        
##  [3] "Root;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae"
##  [4] "Root;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rickettsiales;f__mitochondria"        
##  [5] "Root;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Rhizobiaceae"          
##  [6] "Root;k__Bacteria;p__Actinobacteria;c__Actinobacteria;o__Actinomycetales;f__Kineosporiaceae"        
##  [7] "Root;k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__Streptophyta;f__"                              
##  [8] "Root;k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhizobiales;f__Bradyrhizobiaceae"     
##  [9] "Root;k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales;f__Pseudomonadaceae"  
## [10] "Root;k__Bacteria;p__Bacteroidetes;c__Sphingobacteriia;o__Sphingobacteriales;f__Sphingobacteriaceae"
</code></pre>
<p>Note that there are some odd aspects to the format that could make it challenging to parse:</p>
<ul>
<li>Some taxa have <b><a href ="appendix--00--glossary.html#taxonomic_ranks_anchor">ranks</a></b> (e.g. âk__Bacteriaâ) and some donât (e.g.âUnassigned&quot; and âRootâ).</li>
<li>Some taxa have ranks, but no names (e.g. âf__â).</li>
</ul>
<p>If we just consider the ranks to be a part of the taxon name, then its pretty easy to parse:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(taxa)
obj &lt;-<span class="st"> </span><span class="kw">parse_tax_data</span>(otu_data,
                      <span class="dt">class_cols =</span> <span class="st">&quot;taxonomy&quot;</span>, <span class="co"># The column in the input table</span>
                      <span class="dt">class_sep =</span> <span class="st">&quot;;&quot;</span>) <span class="co"># What each taxon is seperated by</span>
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   1558 taxa: <span style='color: #00BB00;'>aab</span><span>. Unassigned, </span><span style='color: #00BB00;'>aac</span><span>. Root ... </span><span style='color: #00BB00;'>chx</span><span>. f__Methanospirillaceae, </span><span style='color: #00BB00;'>chy</span><span>. f__
##   1558 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aab</span><span>, </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aac</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aae</span><span> ... </span><span style='color: #00BB00;'>bel</span><span>-&gt;</span><span style='color: #00BB00;'>chw</span><span>, </span><span style='color: #00BB00;'>ays</span><span>-&gt;</span><span style='color: #00BB00;'>chx</span><span>, </span><span style='color: #00BB00;'>bem</span><span>-&gt;</span><span style='color: #00BB00;'>chy</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 47,806 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aab</span><span>      1               0        0        0        0        0        0        0
##       2 </span><span style='color: #00BB00;'>ben</span><span>      2              41       22        4      726   112492        2      413
##       3 </span><span style='color: #00BB00;'>beo</span><span>      3              67       65       12    13514        1        4    13314
##       # ... with 4.78e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, â¦
##   0 functions:
</span></code></pre>
<p>Above is the print out of a <code>taxmap</code> object. The first line tells us that the OTUs are assigned to 1,558 unique taxa and lists their IDs and names. These taxon IDs were generated automatically when converting to the <code>taxmap</code> format and were not in the original data set. The second line describes how taxa relate to each other in the tree. Note how our original data are now inside this object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(obj<span class="op">$</span>data<span class="op">$</span>tax_data)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 47,806 x 1,707</span><span>
##    taxon_id OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38 M1551P90
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span>
## </span><span style='color: #BCBCBC;'> 1</span><span> aab      1               0        0        0        0        0        0        0        0
## </span><span style='color: #BCBCBC;'> 2</span><span> ben      2              41       22        4      726   </span><span style='text-decoration: underline;'>112</span><span>492        2      413        2
## </span><span style='color: #BCBCBC;'> 3</span><span> beo      3              67       65       12    </span><span style='text-decoration: underline;'>13</span><span>514        1        4    </span><span style='text-decoration: underline;'>13</span><span>314       70
## </span><span style='color: #BCBCBC;'> 4</span><span> ben      4            </span><span style='text-decoration: underline;'>3</span><span>229    </span><span style='text-decoration: underline;'>13</span><span>679     </span><span style='text-decoration: underline;'>1</span><span>832      951      113     </span><span style='text-decoration: underline;'>2</span><span>496      567     </span><span style='text-decoration: underline;'>2</span><span>428
## </span><span style='color: #BCBCBC;'> 5</span><span> bep      5            </span><span style='text-decoration: underline;'>1</span><span>200       92     </span><span style='text-decoration: underline;'>3</span><span>530     </span><span style='text-decoration: underline;'>2</span><span>008        0      183     </span><span style='text-decoration: underline;'>2</span><span>087      292
## </span><span style='color: #BCBCBC;'> 6</span><span> beq      6             219     </span><span style='text-decoration: underline;'>1</span><span>980     </span><span style='text-decoration: underline;'>1</span><span>200      499        1      781      214     </span><span style='text-decoration: underline;'>2</span><span>273
## </span><span style='color: #BCBCBC;'> 7</span><span> ber      7             485     </span><span style='text-decoration: underline;'>5</span><span>123      755     </span><span style='text-decoration: underline;'>4</span><span>080      443     </span><span style='text-decoration: underline;'>1</span><span>278     </span><span style='text-decoration: underline;'>2</span><span>193      401
## </span><span style='color: #BCBCBC;'> 8</span><span> bes      8             840     </span><span style='text-decoration: underline;'>7</span><span>079     </span><span style='text-decoration: underline;'>3</span><span>760       22        0     </span><span style='text-decoration: underline;'>2</span><span>699       22     </span><span style='text-decoration: underline;'>2</span><span>870
## </span><span style='color: #BCBCBC;'> 9</span><span> bet      9              40       82       91      881        1      449     </span><span style='text-decoration: underline;'>1</span><span>121       90
## </span><span style='color: #BCBCBC;'>10</span><span> beu      10             11       79      277     </span><span style='text-decoration: underline;'>2</span><span>879        1        0     </span><span style='text-decoration: underline;'>6</span><span>811       14
## </span><span style='color: #949494;'># ... with 47,796 more rows, and 1,697 more variables: M1551P71 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P12 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1551P84 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P48 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P4 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P52 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P3 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P15 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>,
## #   M1551P31 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, M1551P75 </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span style='color: #949494;'>, â¦</span><span>
</span></code></pre>
<p><code>obj$data</code> is a list of arbitrary, user-defined data sets. These data sets can be named anything and can be any R object, such as <b><a href ="appendix--00--glossary.html#list_anchor">lists</a></b>, or <b><a href ="appendix--00--glossary.html#vector_anchor">vectors</a></b>, or tables. This is different than <code>phyloseq</code> objects, which have a fixed number of data sets in pre-defined formats, since the focus of <code>taxa</code> is taxonomic data in general and the focus of <code>phyloseq</code> is microbiome data in particular. Note that our data set now has a âtaxon_idâ column, which associates rows in the table to a taxa in the taxonomy. This column is essential for the manipulation functions of <code>taxa</code> to know how to handle these data sets, as we will demonstrate later.</p>
<p>If we want to split out the rank information while parsing, we can use <b><a href ="appendix--00--glossary.html#regular_expressions_anchor">regular expressions</a></b> (a.k.a âregexâ) to specify which part of each taxon is a rank and which part is a name. If you are not familiar with using regular expressions, this might be challenging to understand at first, but it is a very useful skill to have, so it is worth learning. Most regular expressions are composed of a series of âwhat to matchâ followed by âhow many times to matchâ. One regular expression that matches the pattern of taxon names is <code>^[a-z]{0,1}_{0,2}.*$</code>. This might look intimidating, but we can break it down into understandable parts:</p>
<ul>
<li>The <code>^</code> and <code>$</code> represent the start and end of the text respectively. If these were not there, then the pattern could match just a part of the text.</li>
<li>The square brackets (e.g. <code>[a-z]</code>) specify a range of characters that can be matched. Likewise the <code>.</code> means match any character.</li>
<li>The contents of the curly braces (e.g. <code>{0,1}</code>) indicate the number of time the preceding pattern can match. Likewise, the <code>*</code> means 0 or more matches. For example the part of the regex <code>^[a-z]{0,1}</code> means âmatch a character <code>a</code> through <code>z</code> that occurs at the start of the string either zero or one timesâ.</li>
<li>Any text that is not a special regex character (e.g., <code>[</code> and <code>.</code>) matches itself, so the <code>_</code> matches a <code>_</code> in the text. To match characters like <code>[</code> in the text you âescapeâ them with <code>\\</code> (e.g., <code>\\[</code>).</li>
</ul>
<p>The whole regex means the following in common English:</p>
<p>âFrom the start of the string, (<code>^</code>) match any character betweenâa&quot; and âzâ (<code>[a-z]</code>) zero or one times (<code>{0,1}</code>) followed by an underscore (<code>_</code>) occurring between zero and 2 times (<code>{0,2}</code>), followed by any character (<code>.</code>) occurring zero or more times (<code>*</code>), followed by the end of the text (<code>$</code>).&quot;</p>
<p>We can add parentheses that specify which parts of the pattern go together; these are called <b><a href ="appendix--00--glossary.html#capture_groups_anchor">capture groups</a></b> in regex jargon. These do not change what will be matched; they just define different parts of the pattern. In this case, we are interested in the taxon rank (matched by <code>([a-z]{0,1})</code>) and the taxon name (matched by <code>(.*)</code>).</p>
<p>The <code>parse_tax_data</code> function from the <code>taxa</code> package uses regular expressions with capture groups to isolate the pieces of information we want. For each capture group in the regular expression (aka âregexâ), a value is given to the <code>class_key</code> option specifying what the group is (e.g.Â taxon name). Putting this all together, we can read the data like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj &lt;-<span class="st"> </span><span class="kw">parse_tax_data</span>(otu_data,
                      <span class="dt">class_cols =</span> <span class="st">&quot;taxonomy&quot;</span>,
                      <span class="dt">class_sep =</span> <span class="st">&quot;;&quot;</span>,
                      <span class="dt">class_regex =</span> <span class="st">&quot;^([a-z]{0,1})_{0,2}(.*)$&quot;</span>,
                      <span class="dt">class_key =</span> <span class="kw">c</span>(<span class="st">&quot;tax_rank&quot;</span> =<span class="st"> &quot;taxon_rank&quot;</span>, <span class="st">&quot;name&quot;</span> =<span class="st"> &quot;taxon_name&quot;</span>))
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   1558 taxa: <span style='color: #00BB00;'>aab</span><span>. Unassigned, </span><span style='color: #00BB00;'>aac</span><span>. Root ... </span><span style='color: #00BB00;'>chx</span><span>. Methanospirillaceae, </span><span style='color: #00BB00;'>chy</span><span>. 
##   1558 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aab</span><span>, </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aac</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aae</span><span> ... </span><span style='color: #00BB00;'>bel</span><span>-&gt;</span><span style='color: #00BB00;'>chw</span><span>, </span><span style='color: #00BB00;'>ays</span><span>-&gt;</span><span style='color: #00BB00;'>chx</span><span>, </span><span style='color: #00BB00;'>bem</span><span>-&gt;</span><span style='color: #00BB00;'>chy</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 47,806 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aab</span><span>      1               0        0        0        0        0        0        0
##       2 </span><span style='color: #00BB00;'>ben</span><span>      2              41       22        4      726   112492        2      413
##       3 </span><span style='color: #00BB00;'>beo</span><span>      3              67       65       12    13514        1        4    13314
##       # ... with 4.78e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, â¦
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 216,417 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank name       regex_match
## </span><span style='font-style: italic;'>        &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;      </span><span>
##       1 </span><span style='color: #00BB00;'>aab</span><span>                1 ""       Unassigned Unassigned 
##       2 </span><span style='color: #00BB00;'>aac</span><span>                2 ""       Root       Root       
##       3 </span><span style='color: #00BB00;'>aad</span><span>                2 k        Bacteria   k__Bacteria
##       # ... with 2.164e+05 more rows
##   0 functions:
</span></code></pre>
<p>Note how the taxon names do not have the rank information anymore:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">taxon_names</span>(obj))</code></pre></div>
<pre class="r-output"><code>##              aab              aac              aad              aae              aaf 
##     "Unassigned"           "Root"       "Bacteria"        "Archaea" "Proteobacteria" 
##              aag 
## "Actinobacteria"
</code></pre>
<p>Instead, the rank information (and any other capture group content) is in a separate data set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>class_data</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 216,417 x 5</span><span>
##    taxon_id input_index tax_rank name                regex_match           
##    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>          </span><span style='color: #949494;font-style: italic;'>&lt;int&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>               </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                 
## </span><span style='color: #BCBCBC;'> 1</span><span> aab                1 </span><span style='color: #949494;'>""</span><span>       Unassigned          Unassigned            
## </span><span style='color: #BCBCBC;'> 2</span><span> aac                2 </span><span style='color: #949494;'>""</span><span>       Root                Root                  
## </span><span style='color: #BCBCBC;'> 3</span><span> aad                2 k        Bacteria            k__Bacteria           
## </span><span style='color: #BCBCBC;'> 4</span><span> aaf                2 p        Proteobacteria      p__Proteobacteria     
## </span><span style='color: #BCBCBC;'> 5</span><span> add                2 c        Alphaproteobacteria c__Alphaproteobacteria
## </span><span style='color: #BCBCBC;'> 6</span><span> amg                2 o        Rickettsiales       o__Rickettsiales      
## </span><span style='color: #BCBCBC;'> 7</span><span> ben                2 f        mitochondria        f__mitochondria       
## </span><span style='color: #BCBCBC;'> 8</span><span> aac                3 </span><span style='color: #949494;'>""</span><span>       Root                Root                  
## </span><span style='color: #BCBCBC;'> 9</span><span> aad                3 k        Bacteria            k__Bacteria           
## </span><span style='color: #BCBCBC;'>10</span><span> aaf                3 p        Proteobacteria      p__Proteobacteria     
## </span><span style='color: #949494;'># ... with 216,407 more rows</span><span>
</span></code></pre>
<p>However, ranks can also be accessed using the <code>taxon_ranks</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">taxon_ranks</span>(obj))</code></pre></div>
<pre class="r-output"><code>## aab aac aad aae aaf aag 
##  ""  "" "k" "k" "p" "p"
</code></pre>
<p>So we donât really need the âclass_dataâ table, so lets get rid of it:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obj<span class="op">$</span>data<span class="op">$</span>class_data &lt;-<span class="st"> </span><span class="ot">NULL</span></code></pre></div>
<p>Lets also rename the âtax_dataâ table to something more informative:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(obj<span class="op">$</span>data) &lt;-<span class="st"> &quot;otu_counts&quot;</span>
<span class="kw">print</span>(obj)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   1558 taxa: <span style='color: #00BB00;'>aab</span><span>. Unassigned, </span><span style='color: #00BB00;'>aac</span><span>. Root ... </span><span style='color: #00BB00;'>chx</span><span>. Methanospirillaceae, </span><span style='color: #00BB00;'>chy</span><span>. 
##   1558 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aab</span><span>, </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>aac</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aad</span><span>, </span><span style='color: #00BB00;'>aac</span><span>-&gt;</span><span style='color: #00BB00;'>aae</span><span> ... </span><span style='color: #00BB00;'>bel</span><span>-&gt;</span><span style='color: #00BB00;'>chw</span><span>, </span><span style='color: #00BB00;'>ays</span><span>-&gt;</span><span style='color: #00BB00;'>chx</span><span>, </span><span style='color: #00BB00;'>bem</span><span>-&gt;</span><span style='color: #00BB00;'>chy</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>otu_counts</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 47,806 x 1,707</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>OTU_ID M1024P1833 M1551P81 M1551P57 M1551P85 M1551P28 M1551P29 M1551P38
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;</span><span>
##       1 </span><span style='color: #00BB00;'>aab</span><span>      1               0        0        0        0        0        0        0
##       2 </span><span style='color: #00BB00;'>ben</span><span>      2              41       22        4      726   112492        2      413
##       3 </span><span style='color: #00BB00;'>beo</span><span>      3              67       65       12    13514        1        4    13314
##       # ... with 4.78e+04 more rows, and 1,698 more variables: M1551P90 &lt;dbl&gt;,
##       #   M1551P71 &lt;dbl&gt;, M1551P12 &lt;dbl&gt;, M1551P84 &lt;dbl&gt;, M1551P48 &lt;dbl&gt;, M1551P4 &lt;dbl&gt;,
##       #   M1551P52 &lt;dbl&gt;, M1551P3 &lt;dbl&gt;, M1551P15 &lt;dbl&gt;, M1551P31 &lt;dbl&gt;, â¦
##   0 functions:
</span></code></pre>
<p>We can name the tables or other info in <code>obj$data</code> whatever we want. <code>obj$data</code> is a standard <code>list</code>, which means any number of things of any type can be put in it.</p>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<div id="reading-tabular-data" class="section level3">
<h3>Reading tabular data</h3>
<p><strong>1)</strong> Take a look at the file <a href="example_data_1.tsv" class="uri">example_data_1.tsv</a> (click to download).</p>
<p><strong>2)</strong> Try reading the file using the base R function <code>read.table</code>. You might need to change some of the options. Type <code>?read.table</code> to see the documentation for this function. Hint: what does the <code>.tsv</code> extension mean?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-20">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-20" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">table_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;example_data_1.tsv&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
</div>
<p><br /></p>
<p><strong>3)</strong> Now try reading the same file using the <code>read_tsv</code> function from the <code>readr</code> package. Type <code>?read_tsv</code> to see the documentation for this function.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-21">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-21" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(readr)
table_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;example_data_1.tsv&quot;</span>)</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   common_name = col_character(),
##   ncbi_seq_id = col_character(),
##   itis_taxon_id = col_double(),
##   my_taxonomy = col_character()
## )</code></pre>
</div>
<p><br /></p>
<p><strong>4)</strong> Compare the two results. What is different about them? Try using the <code>str</code> function on each result to see the details about how they are formatted.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-22">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-22" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(table_<span class="dv">1</span>)</code></pre></div>
<pre class="r-output"><code>## 'data.frame':    3 obs. of  4 variables:
##  $ common_name  : Factor w/ 3 levels "Bears","Lions",..: 2 3 1
##  $ ncbi_seq_id  : Factor w/ 3 levels "FJ712037","HW243304",..: 1 3 2
##  $ itis_taxon_id: int  183803 183806 180544
##  $ my_taxonomy  : Factor w/ 3 levels "mammalia, felidae, panthera, leo",..: 1 2 3
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(table_<span class="dv">2</span>)</code></pre></div>
<pre class="r-output"><code>## Classes 'tbl_df', 'tbl' and 'data.frame':    3 obs. of  4 variables:
##  $ common_name  : chr  "Lions" "Tigers" "Bears"
##  $ ncbi_seq_id  : chr  "FJ712037" "KC879292" "HW243304"
##  $ itis_taxon_id: num  183803 183806 180544
##  $ my_taxonomy  : chr  "mammalia, felidae, panthera, leo" "mammalia, felidae, panthera, tigris" "mammalia, ursidae, ursus, americanus"
##  - attr(*, "spec")=
##   .. cols(
##   ..   common_name = col_character(),
##   ..   ncbi_seq_id = col_character(),
##   ..   itis_taxon_id = col_double(),
##   ..   my_taxonomy = col_character()
##   .. )
</code></pre>
</div>
<p><br /></p>
<p><strong>5)</strong> Try to change the input parameters of both functions so that all columns are of type <code>character</code> (âchrâ in the <code>str</code> output).</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-23">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-23" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">table_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;example_data_1.tsv&quot;</span>,
                      <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">colClasses =</span> <span class="st">&quot;character&quot;</span>)
table_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;example_data_1.tsv&quot;</span>, <span class="dt">col_types =</span> <span class="st">&quot;cccc&quot;</span>)
<span class="kw">str</span>(table_<span class="dv">1</span>)</code></pre></div>
<pre class="r-output"><code>## 'data.frame':    3 obs. of  4 variables:
##  $ common_name  : chr  "Lions" "Tigers" "Bears"
##  $ ncbi_seq_id  : chr  "FJ712037" "KC879292" "HW243304"
##  $ itis_taxon_id: chr  "183803" "183806" "180544"
##  $ my_taxonomy  : chr  "mammalia, felidae, panthera, leo" "mammalia, felidae, panthera, tigris" "mammalia, ursidae, ursus, americanus"
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(table_<span class="dv">2</span>)</code></pre></div>
<pre class="r-output"><code>## Classes 'tbl_df', 'tbl' and 'data.frame':    3 obs. of  4 variables:
##  $ common_name  : chr  "Lions" "Tigers" "Bears"
##  $ ncbi_seq_id  : chr  "FJ712037" "KC879292" "HW243304"
##  $ itis_taxon_id: chr  "183803" "183806" "180544"
##  $ my_taxonomy  : chr  "mammalia, felidae, panthera, leo" "mammalia, felidae, panthera, tigris" "mammalia, ursidae, ursus, americanus"
##  - attr(*, "spec")=
##   .. cols(
##   ..   common_name = col_character(),
##   ..   ncbi_seq_id = col_character(),
##   ..   itis_taxon_id = col_character(),
##   ..   my_taxonomy = col_character()
##   .. )
</code></pre>
</div>
<p><br /></p>
</div>
<div id="reading-taxonomic-data" class="section level3">
<h3>Reading taxonomic data</h3>
<p>In this exercise we will be converting the table from the previous exercise to a <code>taxmap</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_data &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;example_data_1.tsv&quot;</span>)</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   common_name = col_character(),
##   ncbi_seq_id = col_character(),
##   itis_taxon_id = col_double(),
##   my_taxonomy = col_character()
## )</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(my_data)</code></pre></div>
<pre class="r-output"><code>## <span style='color: #949494;'># A tibble: 3 x 4</span><span>
##   common_name ncbi_seq_id itis_taxon_id my_taxonomy                         
##   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>       </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>               </span><span style='color: #949494;font-style: italic;'>&lt;dbl&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                               
## </span><span style='color: #BCBCBC;'>1</span><span> Lions       FJ712037           </span><span style='text-decoration: underline;'>183</span><span>803 mammalia, felidae, panthera, leo    
## </span><span style='color: #BCBCBC;'>2</span><span> Tigers      KC879292           </span><span style='text-decoration: underline;'>183</span><span>806 mammalia, felidae, panthera, tigris 
## </span><span style='color: #BCBCBC;'>3</span><span> Bears       HW243304           </span><span style='text-decoration: underline;'>180</span><span>544 mammalia, ursidae, ursus, americanus
</span></code></pre>
<p><strong>6)</strong> Look at the documentation for <code>parse_tax_data</code> and <code>lookup_tax_data</code>. The examples at the bottom of the documentation should be helpful.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-25">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-25" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(taxa)
?parse_tax_data
?lookup_tax_data</code></pre></div>
</div>
<p><br /></p>
<p><strong>7)</strong> Try using <code>parse_tax_data</code> to convert the table to a <code>taxmap</code> object, using the <code>my_taxonomy</code> column for the taxonomic information. You should get a taxonomy with âmammaliaâ at the root. How does the table included in the output differ from the input table?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-26">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-26" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">parse_tax_data</span>(my_data, <span class="dt">class_cols =</span> <span class="st">&quot;my_taxonomy&quot;</span>, <span class="dt">class_sep =</span> <span class="st">&quot;, &quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   8 taxa: <span style='color: #00BB00;'>b</span><span>. mammalia, </span><span style='color: #00BB00;'>c</span><span>. felidae, </span><span style='color: #00BB00;'>d</span><span>. ursidae ... </span><span style='color: #00BB00;'>g</span><span>. leo, </span><span style='color: #00BB00;'>h</span><span>. tigris, </span><span style='color: #00BB00;'>i</span><span>. americanus
##   8 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>b</span><span>, </span><span style='color: #00BB00;'>b</span><span>-&gt;</span><span style='color: #00BB00;'>c</span><span>, </span><span style='color: #00BB00;'>b</span><span>-&gt;</span><span style='color: #00BB00;'>d</span><span>, </span><span style='color: #00BB00;'>c</span><span>-&gt;</span><span style='color: #00BB00;'>e</span><span>, </span><span style='color: #00BB00;'>d</span><span>-&gt;</span><span style='color: #00BB00;'>f</span><span>, </span><span style='color: #00BB00;'>e</span><span>-&gt;</span><span style='color: #00BB00;'>g</span><span>, </span><span style='color: #00BB00;'>e</span><span>-&gt;</span><span style='color: #00BB00;'>h</span><span>, </span><span style='color: #00BB00;'>f</span><span>-&gt;</span><span style='color: #00BB00;'>i</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 3 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>common_name ncbi_seq_id itis_taxon_id my_taxonomy                       
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;                             </span><span>
##       1 </span><span style='color: #00BB00;'>g</span><span>        Lions       FJ712037           183803 mammalia, felidae, panthera, leo  
##       2 </span><span style='color: #00BB00;'>h</span><span>        Tigers      KC879292           183806 mammalia, felidae, panthera, tigrâ¦
##       3 </span><span style='color: #00BB00;'>i</span><span>        Bears       HW243304           180544 mammalia, ursidae, ursus, americaâ¦
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>8)</strong> Try using <code>lookup_tax_data</code> to convert the table to a <code>taxmap</code> object, using the <code>ncbi_seq_id</code> column to look up the taxonomy associated with these Genbank accession numbers. NOTE: This requires an internet connection.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-27">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-27" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lookup_tax_data</span>(my_data, <span class="dt">type =</span> <span class="st">&quot;seq_id&quot;</span>, <span class="dt">column =</span> <span class="st">&quot;ncbi_seq_id&quot;</span>, <span class="dt">database =</span> <span class="st">&quot;ncbi&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   33 taxa: <span style='color: #00BB00;'>131567</span><span>. cellular organisms, </span><span style='color: #00BB00;'>2759</span><span>. Eukaryota ... </span><span style='color: #00BB00;'>9694</span><span>. Panthera tigris
##   33 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>131567</span><span>, </span><span style='color: #00BB00;'>131567</span><span>-&gt;</span><span style='color: #00BB00;'>2759</span><span>, </span><span style='color: #00BB00;'>2759</span><span>-&gt;</span><span style='color: #00BB00;'>33154</span><span> ... </span><span style='color: #00BB00;'>9688</span><span>-&gt;</span><span style='color: #00BB00;'>9689</span><span>, </span><span style='color: #00BB00;'>9688</span><span>-&gt;</span><span style='color: #00BB00;'>9694</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 33 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>ncbi_name          ncbi_rank    ncbi_id
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;              &lt;chr&gt;        &lt;chr&gt;  </span><span>
##       1 </span><span style='color: #00BB00;'>131567</span><span>   cellular organisms no rank      131567 
##       2 </span><span style='color: #00BB00;'>2759</span><span>     Eukaryota          superkingdom 2759   
##       3 </span><span style='color: #00BB00;'>33154</span><span>    Opisthokonta       no rank      33154  
##       # ... with 30 more rows
##     </span><span style='font-weight: bold;'>query_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 3 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>common_name ncbi_seq_id itis_taxon_id my_taxonomy                       
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;                             </span><span>
##       1 </span><span style='color: #00BB00;'>9689</span><span>     Lions       FJ712037           183803 mammalia, felidae, panthera, leo  
##       2 </span><span style='color: #00BB00;'>9694</span><span>     Tigers      KC879292           183806 mammalia, felidae, panthera, tigrâ¦
##       3 </span><span style='color: #00BB00;'>9643</span><span>     Bears       HW243304           180544 mammalia, ursidae, ursus, americaâ¦
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>9)</strong> Try using <code>lookup_tax_data</code> to convert the table to a <code>taxmap</code> object, using the <code>itis_taxon_id</code> column to look up the taxonomic classifications from the Integrated Taxonomic Information System (ITIS) for these taxon IDs. NOTE: This requires an internet connection.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-28">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-28" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lookup_tax_data</span>(my_data, <span class="dt">type =</span> <span class="st">&quot;taxon_id&quot;</span>, <span class="dt">column =</span> <span class="st">&quot;itis_taxon_id&quot;</span>, <span class="dt">database =</span> <span class="st">&quot;itis&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   22 taxa: <span style='color: #00BB00;'>202423</span><span>. Animalia, </span><span style='color: #00BB00;'>914154</span><span>. Bilateria ... </span><span style='color: #00BB00;'>183806</span><span>. Panthera tigris tigris
##   22 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>202423</span><span>, </span><span style='color: #00BB00;'>202423</span><span>-&gt;</span><span style='color: #00BB00;'>914154</span><span> ... </span><span style='color: #00BB00;'>180592</span><span>-&gt;</span><span style='color: #00BB00;'>183805</span><span>, </span><span style='color: #00BB00;'>183805</span><span>-&gt;</span><span style='color: #00BB00;'>183806</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 22 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>itis_name     itis_rank    itis_id
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;         &lt;chr&gt;        &lt;chr&gt;  </span><span>
##       1 </span><span style='color: #00BB00;'>202423</span><span>   Animalia      kingdom      202423 
##       2 </span><span style='color: #00BB00;'>914154</span><span>   Bilateria     subkingdom   914154 
##       3 </span><span style='color: #00BB00;'>914156</span><span>   Deuterostomia infrakingdom 914156 
##       # ... with 19 more rows
##     </span><span style='font-weight: bold;'>query_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 3 x 5</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>common_name ncbi_seq_id itis_taxon_id my_taxonomy                       
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;                             </span><span>
##       1 </span><span style='color: #00BB00;'>183803</span><span>   Lions       FJ712037           183803 mammalia, felidae, panthera, leo  
##       2 </span><span style='color: #00BB00;'>183806</span><span>   Tigers      KC879292           183806 mammalia, felidae, panthera, tigrâ¦
##       3 </span><span style='color: #00BB00;'>180544</span><span>   Bears       HW243304           180544 mammalia, ursidae, ursus, americaâ¦
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>10)</strong> Compare the results of the three sources of taxonomic information. What is different? What is the same?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-29">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-29" class="collapse">
<ul>
<li>The number and name of taxa are different since each source has a different taxonomy.</li>
<li>The taxon IDs are different. When an online database is queried, the taxon IDs from that database are used.</li>
<li>The input data stored in âtax_dataâ is the same, except for the taxon ID column.
</div>
<br /></li>
</ul>
</div>
<div id="reading-taxonomic-data-from-complex-formats" class="section level3">
<h3>Reading taxonomic data from complex formats</h3>
<p>Sometimes taxonomic data can be embedded in complex text, like FASTA headers. Look at the file <a href="example_data_2.fa" class="uri">example_data_2.fa</a> (click to download) which contains headers with two sources of taxonomic information:</p>
<ul>
<li>The Genbank accession number</li>
<li>The taxonomic classification</li>
</ul>
<p><strong>11)</strong> Read the FASTA file <a href="example_data_2.fa" class="uri">example_data_2.fa</a> using the <code>read.FASTA</code> function from the <code>ape</code> package and store the result in a variable.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-30">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-30" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ape)</code></pre></div>
<pre><code>## 
## Attaching package: &#39;ape&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:metacoder&#39;:
## 
##     complement</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">seqs &lt;-<span class="st"> </span><span class="kw">read.FASTA</span>(<span class="st">&quot;example_data_2.fa&quot;</span>)
<span class="kw">print</span>(seqs)</code></pre></div>
<pre class="r-output"><code>## 3 DNA sequences in binary format stored in a list.
## 
## Mean sequence length: 455.667 
##    Shortest sequence: 354 
##     Longest sequence: 509 
## 
## Labels:
## FJ712037::class_mammalia;family_felidae;genus_panthera;speci...
## KC879292::class_mammalia;family_felidae;genus_panthera;speci...
## HW243304::class_mammalia;family_ursidae;genus_ursus;species_...
## 
## Base composition:
##     a     c     g     t 
## 0.282 0.154 0.260 0.304
</code></pre>
</div>
<p><br /></p>
<p><strong>12)</strong> Use the <code>names</code> function to get the headers and stores those in another variable.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-31">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-31" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">headers &lt;-<span class="st"> </span><span class="kw">names</span>(seqs)
<span class="kw">print</span>(headers)</code></pre></div>
<pre class="r-output"><code>## [1] "FJ712037::class_mammalia;family_felidae;genus_panthera;species_leo"    
## [2] "KC879292::class_mammalia;family_felidae;genus_panthera;species_tigris" 
## [3] "HW243304::class_mammalia;family_ursidae;genus_ursus;species_americanus"
</code></pre>
</div>
<p><br /></p>
<p><strong>13)</strong> Look at the documentation for <code>extract_tax_data</code> from the <code>taxa</code> package by typing <code>?extract_tax_data</code>.</p>
<p><strong>14)</strong> Use the <code>extract_tax_data</code> function to convert the headers to a <code>taxmap</code> object, using the NCBI accession number. The classification can be ignored (i.e.Â not given a capture group and key value) or stored as <code>&quot;info&quot;</code>.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-32">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-32" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">extract_tax_data</span>(headers,
                 <span class="dt">key =</span> <span class="kw">c</span>(<span class="dt">my_acc_no =</span> <span class="st">&quot;seq_id&quot;</span>, <span class="dt">my_class =</span> <span class="st">&quot;info&quot;</span>),
                 <span class="dt">regex =</span> <span class="st">&quot;^(.+)::(.+)$&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   33 taxa: <span style='color: #00BB00;'>131567</span><span>. cellular organisms, </span><span style='color: #00BB00;'>2759</span><span>. Eukaryota ... </span><span style='color: #00BB00;'>9694</span><span>. Panthera tigris
##   33 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>131567</span><span>, </span><span style='color: #00BB00;'>131567</span><span>-&gt;</span><span style='color: #00BB00;'>2759</span><span>, </span><span style='color: #00BB00;'>2759</span><span>-&gt;</span><span style='color: #00BB00;'>33154</span><span> ... </span><span style='color: #00BB00;'>9688</span><span>-&gt;</span><span style='color: #00BB00;'>9689</span><span>, </span><span style='color: #00BB00;'>9688</span><span>-&gt;</span><span style='color: #00BB00;'>9694</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 33 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>ncbi_name          ncbi_rank    ncbi_id
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;              &lt;chr&gt;        &lt;chr&gt;  </span><span>
##       1 </span><span style='color: #00BB00;'>131567</span><span>   cellular organisms no rank      131567 
##       2 </span><span style='color: #00BB00;'>2759</span><span>     Eukaryota          superkingdom 2759   
##       3 </span><span style='color: #00BB00;'>33154</span><span>    Opisthokonta       no rank      33154  
##       # ... with 30 more rows
##     </span><span style='font-weight: bold;'>query_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 3 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>my_acc_no my_class                       input                          
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;                          &lt;chr&gt;                          </span><span>
##       1 </span><span style='color: #00BB00;'>9689</span><span>     FJ712037  class_mammalia;family_felidaeâ¦ FJ712037::class_mammalia;familâ¦
##       2 </span><span style='color: #00BB00;'>9694</span><span>     KC879292  class_mammalia;family_felidaeâ¦ KC879292::class_mammalia;familâ¦
##       3 </span><span style='color: #00BB00;'>9643</span><span>     HW243304  class_mammalia;family_ursidaeâ¦ HW243304::class_mammalia;familâ¦
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>15)</strong> Now use the classification instead of the accession numbers. The accession number can be ignored or stored as <code>&quot;info&quot;</code>. Consider the rank as part of the taxon name for now.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-33">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-33" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">extract_tax_data</span>(headers,
                 <span class="dt">key =</span> <span class="kw">c</span>(<span class="dt">my_acc_no =</span> <span class="st">&quot;info&quot;</span>, <span class="dt">my_class =</span> <span class="st">&quot;class&quot;</span>),
                 <span class="dt">regex =</span> <span class="st">&quot;^(.+)::(.+)$&quot;</span>,
                 <span class="dt">class_sep =</span> <span class="st">&quot;;&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   8 taxa: <span style='color: #00BB00;'>b</span><span>. class_mammalia, </span><span style='color: #00BB00;'>c</span><span>. family_felidae ... </span><span style='color: #00BB00;'>i</span><span>. species_americanus
##   8 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>b</span><span>, </span><span style='color: #00BB00;'>b</span><span>-&gt;</span><span style='color: #00BB00;'>c</span><span>, </span><span style='color: #00BB00;'>b</span><span>-&gt;</span><span style='color: #00BB00;'>d</span><span>, </span><span style='color: #00BB00;'>c</span><span>-&gt;</span><span style='color: #00BB00;'>e</span><span>, </span><span style='color: #00BB00;'>d</span><span>-&gt;</span><span style='color: #00BB00;'>f</span><span>, </span><span style='color: #00BB00;'>e</span><span>-&gt;</span><span style='color: #00BB00;'>g</span><span>, </span><span style='color: #00BB00;'>e</span><span>-&gt;</span><span style='color: #00BB00;'>h</span><span>, </span><span style='color: #00BB00;'>f</span><span>-&gt;</span><span style='color: #00BB00;'>i</span><span>
##   1 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 3 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>my_acc_no my_class                       input                          
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;                          &lt;chr&gt;                          </span><span>
##       1 </span><span style='color: #00BB00;'>g</span><span>        FJ712037  class_mammalia;family_felidaeâ¦ FJ712037::class_mammalia;familâ¦
##       2 </span><span style='color: #00BB00;'>h</span><span>        KC879292  class_mammalia;family_felidaeâ¦ KC879292::class_mammalia;familâ¦
##       3 </span><span style='color: #00BB00;'>i</span><span>        HW243304  class_mammalia;family_ursidaeâ¦ HW243304::class_mammalia;familâ¦
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>16)</strong> Now try to split out the rank information from the taxon name. You will need to use the <code>class_key</code> and <code>class_regex</code> options. The ranks can be stored as <code>&quot;info&quot;</code> or <code>&quot;taxon_rank</code>â. Using <code>&quot;taxon_rank</code>â allows the ranks to be accessed with the <code>taxon_ranks</code> function.</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-34">
Show solution
</button>
<div id="hide_buttonunnamed-chunk-34" class="collapse">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">extract_tax_data</span>(headers,
                 <span class="dt">key =</span> <span class="kw">c</span>(<span class="dt">acc_no =</span> <span class="st">&quot;info&quot;</span>, <span class="dt">my_class =</span> <span class="st">&quot;class&quot;</span>),
                 <span class="dt">regex =</span> <span class="st">&quot;^(.+)::(.+)$&quot;</span>,
                 <span class="dt">class_sep =</span> <span class="st">&quot;;&quot;</span>,
                 <span class="dt">class_key =</span> <span class="kw">c</span>(<span class="dt">tax_rank =</span> <span class="st">&quot;info&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;taxon_name&quot;</span>),
                 <span class="dt">class_regex =</span> <span class="st">&quot;^(.+)_(.+)$&quot;</span>)</code></pre></div>
<pre class="r-output"><code>## &lt;Taxmap&gt;
##   8 taxa: <span style='color: #00BB00;'>b</span><span>. mammalia, </span><span style='color: #00BB00;'>c</span><span>. felidae, </span><span style='color: #00BB00;'>d</span><span>. ursidae ... </span><span style='color: #00BB00;'>g</span><span>. leo, </span><span style='color: #00BB00;'>h</span><span>. tigris, </span><span style='color: #00BB00;'>i</span><span>. americanus
##   8 edges: </span><span style='color: #00BB00;'>NA</span><span>-&gt;</span><span style='color: #00BB00;'>b</span><span>, </span><span style='color: #00BB00;'>b</span><span>-&gt;</span><span style='color: #00BB00;'>c</span><span>, </span><span style='color: #00BB00;'>b</span><span>-&gt;</span><span style='color: #00BB00;'>d</span><span>, </span><span style='color: #00BB00;'>c</span><span>-&gt;</span><span style='color: #00BB00;'>e</span><span>, </span><span style='color: #00BB00;'>d</span><span>-&gt;</span><span style='color: #00BB00;'>f</span><span>, </span><span style='color: #00BB00;'>e</span><span>-&gt;</span><span style='color: #00BB00;'>g</span><span>, </span><span style='color: #00BB00;'>e</span><span>-&gt;</span><span style='color: #00BB00;'>h</span><span>, </span><span style='color: #00BB00;'>f</span><span>-&gt;</span><span style='color: #00BB00;'>i</span><span>
##   2 data sets:
##     </span><span style='font-weight: bold;'>tax_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 3 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>acc_no  my_class                        input                           
## </span><span style='font-style: italic;'>        &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;                           &lt;chr&gt;                           </span><span>
##       1 </span><span style='color: #00BB00;'>g</span><span>        FJ7120â¦ class_mammalia;family_felidae;â¦ FJ712037::class_mammalia;familyâ¦
##       2 </span><span style='color: #00BB00;'>h</span><span>        KC8792â¦ class_mammalia;family_felidae;â¦ KC879292::class_mammalia;familyâ¦
##       3 </span><span style='color: #00BB00;'>i</span><span>        HW2433â¦ class_mammalia;family_ursidae;â¦ HW243304::class_mammalia;familyâ¦
##     </span><span style='font-weight: bold;'>class_data</span><span>:
## </span><span style='font-style: italic;'>      # A tibble: 12 x 4</span><span>
##        </span><span style='color: #00BB00;'> taxon_id </span><span>input_index tax_rank name    
## </span><span style='font-style: italic;'>      * &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;   </span><span>
##       1 </span><span style='color: #00BB00;'>b</span><span>                  1 class    mammalia
##       2 </span><span style='color: #00BB00;'>c</span><span>                  1 family   felidae 
##       3 </span><span style='color: #00BB00;'>e</span><span>                  1 genus    panthera
##       # ... with 9 more rows
##   0 functions:
</span></code></pre>
</div>
<p><br /></p>
<p><strong>17)</strong> Sometimes you will have the option to use more than one source of taxonomic information, like in this exercise. What are the benefits of using the embedded classification information instead of a Genbank accession number or taxon ID?</p>
<button class="btn btn-danger" data-toggle="collapse" data-target="#hide_buttonunnamed-chunk-35">
Show Answer
</button>
<div id="hide_buttonunnamed-chunk-35" class="collapse">
Using the included classification does not require communicating with online databases, so is much faster and more reliable.
</div>
<p><br /></p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-wagner2016host">
<p>Wagner, Maggie R, Derek S Lundberg, G Tijana, Susannah G Tringe, Jeffery L Dangl, and Thomas Mitchell-Olds. 2016. âHost Genotype and Age Shape the Leaf and Root Microbiomes of a Wild Perennial Plant.â <em>Nature Communications</em> 7. Nature Publishing Group: 12151. doi:<a href="https://doi.org/10.1038/ncomms12151">10.1038/ncomms12151</a>.</p>
</div>
<div id="ref-zeeberg2004mistaken">
<p>Zeeberg, Barry R, Joseph Riss, David W Kane, Kimberly J Bussey, Edward Uchio, W Marston Linehan, J Carl Barrett, and John N Weinstein. 2004. âMistaken Identifiers: Gene Name Errors Can Be Introduced Inadvertently When Using Excel in Bioinformatics.â <em>BMC Bioinformatics</em> 5 (1). BioMed Central: 80. <a href="https://doi.org/10.1186/1471-2105-5-80" class="uri">https://doi.org/10.1186/1471-2105-5-80</a>.</p>
</div>
</div>
</div>
</div>

<div class="footer_section">
  <div class="footer_content">
    <span xmlns:dct="http://purl.org/dc/terms/" href="http://purl.org/dc/dcmitype/Text" property="dct:title" rel="dct:type">Metacoder's documentation</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://grunwaldlab.cgrb.oregonstate.edu/" property="cc:attributionName" rel="cc:attributionURL">The Grunwald lab and the USDA Horticultural Crops Research Unit</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.<br />Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/grunwaldlab/metacoder_documentation" rel="dct:source">https://github.com/grunwaldlab/metacoder_documentation</a>.
  </div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
